---
layout: post
title: "MA plot to visualize gene expression data using Python"
date:   2018-12-09 010:19:20
author: Renesh Bedre
description: "Advanced Bioinformatics"
permalink: blog/ma.html
comments: true
---

**<span style="color:#33a8ff">What is MA plot?</span>**
 - 2-dimensional (2D) scatter plot to visualize gene expression datasets
 - Visualize and identify  gene expression changes from two different conditions (eg. normal vs. 
   treated) in terms of log fold change (M) on Y-axis and log of mean of expression counts of normal and treated 
   samples (A) on X-axis
 - Genes with similar expression values in both normal and treated samples will cluster around M=0 value i.e genes 
   expressed with no significant differences in between treatments
 - Points away from M=0 line indicates genes with significant expression, For example, gene is upregulated and 
   downregulated if point is above and below M=0 line respectively
 - MA plot does not consider statistical measures (P-values or adjusted P-values) and therefore we can not tell genes 
   with statistically significant differences between normal vs. treated from MA plot (Use 
   <a href="https://reneshbedre.github.io/blog/volcano.html" target="_blank">Volcano plot</a> if you want indicates 
   genes with statistically significant differences)
   
**<span style="color:#33a8ff">How to create MA plot in Python?</span>**

For generating MA plot, I have used gene expression data published in Bedre et al. 2016 to identify significantly
induced or downregulated genes in response to salt stress in <i>Spartina alterniflora</i> 
(<a href="https://bmcgenomics.biomedcentral.com/articles/10.1186/s12864-016-3017-3">Read paper</a>). 
Here's you can download gene expression dataset used for generating MA plot: 
[dataset]({{"/myfiles/volcano/SaLR_DEGseq.txt" target="_blank""| absolute_url }})

```python
# I am using Python3
# load required packages
# make sure you have installed required packages
import pandas as pd
import matplotlib.pyplot as plt
import numpy as np

# load data file
d = pd.read_csv("https://reneshbedre.github.io/myfiles/ma/SaLR_DEGseq.csv")

# to see first few lines of table and get table dimensions
print(d.head()) # first five lines
print(d.info()) # table information including dimensions

#  this gene expression dataset has 19025 genes
# we need to add color column to visualize up, downregulated and intermediate genes
# here you can also change the log2 fold change (log2FC) numbers 
# as per your need
d.loc[(d['log2FC'] >= 1), 'color'] = "green"  # upregulated
d.loc[(d['log2FC'] <=- 1), 'color'] = "red"   # downregulated
d['color'].fillna('grey', inplace=True) # intermediate

# to reduce the noise, filter out genes with low expression counts across treatments 
# (say, < 10 normalized expression count)
# you can change this number as per your requirement and based on expression unit 
# value1 and value2 represents counts for stress and control treatments respectively
d = d.loc[(d['value1'] >= 10) & (d['value2'] >= 10)]

# Now, data is ready for MA plot
# In MA plot, X-axis is log2 normalized mean of expression counts
d['A']=np.log2( (d['value1']+d['value2'])/2 )

# plot 
plt.scatter(d['A'], d['log2FC'], c=d['color'])
# draw a central line at M=0
plt.axhline(y=0, color='b', linestyle='--')
plt.xlabel('A',fontsize=15, fontname="sans-serif", fontweight="bold")
plt.ylabel('M', fontsize=15, fontname="sans-serif", fontweight="bold")
plt.xticks(fontsize=12, fontname="sans-serif")
plt.yticks(fontsize=12, fontname="sans-serif")
plt.show()
# To save MA plot to file, replace  plt.show() with following line
plt.savefig('MA.png', format='png', bbox_inches='tight', dpi=300)
```
Generated MA plot by adding above code (green: upregulated and red: downregulated genes),

![screenshot]({{ "/myfiles/ma/MA.png" | absolute_url }})

If you would like to add specific gene names in MA plot, use following code,

```python
# plot 
plt.scatter(d['A'], d['log2FC'], c=d['color'])
# draw a central line at M=0
plt.axhline(y=0, color='b', linestyle='--')
plt.xlabel('A',fontsize=15, fontname="sans-serif", fontweight="bold")
plt.ylabel('M', fontsize=15, fontname="sans-serif", fontweight="bold")
plt.xticks(fontsize=12, fontname="sans-serif")
plt.yticks(fontsize=12, fontname="sans-serif")
# I have added two gene names. You can add multiple gene names to corresponding point  
# using axis coordinates
plt.text(7.34, 4.09, "CPuORF26")
plt.text(8.06,-2.23, "CIA")
plt.show()
# To save MA plot to file, replace  plt.show() with following line
plt.savefig('MA_text.png', format='png', bbox_inches='tight', dpi=300)
```

Generated volcano plot by adding above code,

![screenshot]({{ "/myfiles/ma/MA_text.png" | absolute_url }})


**<span style="color:#33a8ff">How to cite?</span>**
Bedre, R. “MA plot to visualize gene expression data using Python” Renesh Bedre (blog), December 9, 2018, 
https://reneshbedre.github.io/blog/ma.html.

<span style="color:#9e9696">If you have any questions, comments or recommendations, please email me at 
<b>reneshbe@gmail.com</b></span>
    