---
title: "MA plot to visualize gene expression data using Python"
date:   2018-12-09 010:19:20
author_profile: true
permalink: blog/ma.html
classes: wide
tags:
  - Genomic visualization
  - Gene expression
---

<p>
{% include  share.html %}
</p>

## <span style="color:#33a8ff">What is MA plot?</span>
 - 2-dimensional (2D) scatter plot to visualize gene expression datasets
 - Visualize and identify  gene expression changes from two different conditions (eg. normal vs. 
   treated) in terms of log fold change (M) on Y-axis and log of the mean of normalized expression counts of normal and treated
   samples (A) on X-axis
 - Genes with similar expression values in both normal and treated samples will cluster around M=0 value i.e genes 
   expressed with no significant differences in between treatments
 - Points away from M=0 line indicate genes with significant expression, For example, a gene is upregulated and 
   downregulated if the point is above and below M=0 line respectively
 - MA plot does not consider statistical measures (P-values or adjusted P-values) and therefore we can not tell genes 
   with statistically significant differences between normal vs. treated from MA plot (Use 
   <a href="https://reneshbedre.github.io/blog/volcano.html" target="_blank">Volcano plot</a> if you want indicates 
   genes with statistically significant differences)
   
## <span style="color:#33a8ff">How to create MA plot in Python?</span>
- We will use `bioinfokit v0.8.7` or later
- Check [bioinfokit documentation]({{"/blog/howtoinstall.html" | absolute_url }}) for installation and documentation
- For generating the MA plot, I have used gene expression data published in Bedre et al. 2016 to identify 
  statistically significantly induced or downregulated genes in response to salt stress in <i>Spartina alterniflora</i> 
  (<a href="https://bmcgenomics.biomedcentral.com/articles/10.1186/s12864-016-3017-3">Read paper</a>). Here's you can 
  download gene expression dataset used for generating MA plot: 
  [dataset]({{"/assets/posts/ma/test_dataset.csv" | absolute_url }})

```python
# you can use interactive python interpreter, jupyter notebook, spyder or python code
# I am using interactive python interpreter (Python 3.7)
>>> from bioinfokit import analys, visuz
# load dataset as pandas dataframe
>>> df = analys.get_data('ma').data
>>> df.head()
          GeneNames  value1  value2    log2FC       p-value
0  LOC_Os09g01000.1    8862   32767 -1.886539  1.250000e-55
1  LOC_Os12g42876.1    1099     117  3.231611  1.050000e-55
2  LOC_Os12g42884.2     797      88  3.179004  2.590000e-54
3  LOC_Os03g16920.1     274       7  5.290677  4.690000e-54
4  LOC_Os05g47540.4     308      18  4.096862  2.190000e-54

>>> visuz.gene_exp.ma(df=df, lfc='log2FC', ct_count='value1', st_count='value2')
# plot will be saved in same directory (ma.png)
# set parameter show=True, if you want view the image instead of saving
```

Generated MA plot by adding above code (green: upregulated and red: downregulated genes),

<p align="center">
<img src="/assets/posts/ma/ma1.svg" width="600">
</p>

Add legend to the plot,
```python
>>> visuz.gene_exp.ma(df=df, lfc='log2FC', ct_count='value1', st_count='value2', plotlegend=True)
```

<p align="center">
<img src="/assets/posts/ma/ma1_1.svg" width="600">
</p>

Change color of MA plot
```python
# change colormap
>>> visuz.gene_exp.ma(df=df, lfc='log2FC', ct_count='value1', st_count='value2', plotlegend=True, 
    color=('#00239CFF', 'grey', '#E10600FF'))
```

<p align="center">
<img src="/assets/posts/ma/ma2.svg" width="600">
</p>

Change log fold change threshold,
```python
>>> visuz.gene_exp.ma(df=df, lfc='log2FC', ct_count='value1', st_count='value2', lfc_thr=2, plotlegend=True, 
    color=('#00239CFF', 'grey', '#E10600FF'))
```

<p align="center">
<img src="/assets/posts/ma/ma2_1.svg" width="600">
</p>

Change the shape of the points
```python
# add star shape
# check more shapes at https://matplotlib.org/3.1.1/api/markers_api.html
>>> visuz.gene_exp.ma(df=df, lfc='log2FC', ct_count='value1', st_count='value2', color=('#00239CFF', 'grey', '#E10600FF'), 
    markerdot='*', plotlegend=True)
```

<p align="center">
<img src="/assets/posts/ma/ma3.svg" width="600">
</p>

Change the transparency of the points
```python
# add star shape
# check more shapes at https://matplotlib.org/3.1.1/api/markers_api.html
>>> visuz.gene_exp.ma(df=df, lfc='log2FC', ct_count='value1', st_count='value2', color=('#00239CFF', 'grey', '#E10600FF'), 
    markerdot='*', valpha=0.5, plotlegend=True)
```

<p align="center">
<img src="/assets/posts/ma/ma4.svg" width="600">
</p>

Draw log fold change threshold lines

```python
# change colormap
>>> visuz.gene_exp.ma(df=df, lfc='log2FC', ct_count='value1',  st_count='value2', color=('#00239CFF', 'grey', '#E10600FF'),
    fclines=True, plotlegend=True)
```

<p align="center">
<img src="/assets/posts/ma/ma41.svg" width="600">
</p>


Change X and Y range ticks, font size and name for tick labels
```python
>>> visuz.gene_exp.ma(df=df, lfc='log2FC', ct_count='value1', st_count='value2', color=('#00239CFF', 'grey', '#E10600FF'), 
    markerdot='*', figtype='svg', xlm=(0,16,1), ylm=(-6,6,1), axtickfontsize=10, axtickfontname='Verdana', plotlegend=True)
```

<p align="center">
<img src="/assets/posts/ma/ma5.svg" width="600">
</p>

Change legend position and labels
```python
>>> visuz.gene_exp.ma(df=df, lfc='log2FC', ct_count='value1', st_count='value2', color=('#00239CFF', 'grey', '#E10600FF'),
    plotlegend=True, legendpos='lower right', legendlabels=['Upregulated', 'Normal', 'Downregulated'])
```

<p align="center">
<img src="/assets/posts/ma/ma6.svg" width="600">
</p>


In addition to these parameters, the parameters for figure type (`figtype`), X and Y axis ticks range (`xlm`, `ylm`), axis labels (`axxlabel`, `axylabel`),  
axis labels font size and name (`axlabelfontsize`, `axlabelfontname`), and axis tick labels font size and name (`axtickfontsize`, `axtickfontname`)
can be provided.

Check detailed <a href='https://reneshbedre.github.io/blog/howtoinstall.html#ma-plot' target='_blank'>usage</a>


<p>
{% include  cite.html %}
</p>

<p>
{% include  share.html %}
</p>
    
<span style="color:#9e9696"><i> Last updated: June 27, 2020</i> </span>

<p>
{% include  license.html %}
</p>