---
title: "Merge multiple VCF files"
date:   2019-03-18 010:19:20
permalink: blog/mergevcf.html
author_profile: true
classes: wide
---

<p>
{% include  share.html %}
</p>

**<span style="color:#33a8ff">What is VCF file?</span>**
- VCF stands for variant call format
- It is a text file (file extension as .vcf) storing meta-information, marker and genotype data of genetic variations

**<span style="color:#33a8ff">How to merge multiple VCF files </span>** 

Sometimes, it is necessary to merge different VCF files for analysis as the genotype information stored in multiple 
files (For example, you have different VCF file for every chromosome).


```python
# We will use `bioinfokit` for merging different VCF files
# download and install bioinfokit (Tested on Linux, Mac, Windows) 
git clone https://github.com/reneshbedre/bioinfokit.git
cd bioinfokit
python3 setup.py install
```

```python
# you can use interactive python console, jupyter or python code
# I am using interactive python console (Python 3.6)
# go to a directory where all vcf files are stored. make sure all files are uncompressed.
# make sure you will have uniform VCF files. For example, multiple VCF files 
# generated from same source datasets
>>>from bioinfokit import analys
# merge VCF files. You can provide multiple VCF files separated by comma.
>>>analys.mergevcf("file_1.vcf,file_2.vcf,file_3.vcf,file_4.vcf")
# merged VCF files will be stored in same directory (merge_vcf.vcf)
```


<p>
{% include  cite.html %}
</p>

<p>
{% include  share.html %}
</p>

<span style="color:#9e9696"><i> Last updated: March 18, 2019</i> </span>
