---
title: "Sequencing coverage and breadth of coverage" 
date: 2019-11-23 010:19:20 
permalink: blog/seqcov.html 
author_profile: true 
classes: wide
tags:
  - High-throughput sequence analysis
---

This page has been moved <a href='https://www.reneshbedre.com/blog/seqcov.html' target='_blank'>here</a>


<!-- 1

## <span style="color:#33a8ff">What is Sequencing coverage or depth and breadth of coverage?</span>
- Sequencing coverage or depth (coverage and depth are used interchangeably) 
  determines the number of times sequenced nucleotide bases covered the target genome. 
  For example, if genome size is 100 Mbp and you have sequenced 5 M 
  reads of 100 bp size, then sequencing coverage at genome level would be 5X.
- The breadth of coverage refers to the percentage of genome bases 
  sequenced at a given sequencing depth. For example, if 95% of the 
  genome is covered by sequencing at a certain depth.

## <span style="color:#33a8ff">How to calculate sequencing coverage? </span>
- Sequencing coverage is calculated based on the type of sequencing. For RNA-seq applications, coverage is
  calculated based on the transcriptome size and for genome sequencing applications, coverage is
  calculated based on the genome size
- Generally in RNA-seq experiments, the read depth (number of reads per sample) is used instead of coverage.
  High read depth is necessary to identify genes with low expressions. The typical read depth RNA-seq 
  experiment to study gene expression ranges from 5 to 25 M reads per sample. 
- Calculating sequencing coverage based on raw sequence reads will give you rough estimates as some of these raw reads may
  have contamination (adapter, primer, duplicates or low-quality bases) or may not map to genome. In such cases, you 
  can consider genome mapped data for estimating the coverage. 

We will use `bioinfokit` (v0.9.7 or later) <br>
Check [bioinfokit documentation]({{"/blog/howtoinstall.html" | absolute_url }}) for installation and documentation

```python
# you can use interactive python interpreter, jupyter notebook, spyder or python code
# I am using interactive python interpreter (Python 3.7)
# go to a directory where fastq files are saved. Make sure fastq file is uncompressed.
# this will give sequencing coverage per sample
>>> from bioinfokit.analys import fastq
>>> fastq.seqcov(file="fastq_file", gs="genome size in Mbp")
```

## <span style="color:#33a8ff">Recommended sequence coverage for sequencing applications</span>

| Sequencing applications  | Recommended Coverage  |
|----|----|
| Whole genome sequencing (WGS) | 15X to 60X |
| Whole exome sequencing (WES) | 100X |
| RNA sequencing (RNA-seq) | 5 to 100 M reads per sample depending on target study |
| ChIP-Seq | 100X |

Source: Illumina and genohub

References:
- Sims D, Sudbery I, Ilott NE, Heger A, Ponting CP. Sequencing depth and 
  coverage: key considerations in genomic analyses. Nature Reviews Genetics. 2014 Feb;15(2):121-32. 

1 -->

<p>
{% include  cite.html %}
</p>

<p>
{% include  share.html %}
</p>

<!-- 1
<span style="color:#9e9696"><i> Last updated: September 18, 2020</i> </span>
1 -->

<p>
{% include  license.html %}
</p>