---
layout: post
title: "Heatmap in Python"
date:   2019-11-26 02:15:18
author: Renesh Bedre
description: "Advanced Bioinformatics"
permalink: blog/hmap.html
comments: true
---
<p>
{% include  share.html %}
</p>

**<span style="color:#33a8ff">What is heatmap?</span>**
 - Continuous colormap where each color represents a specific set of values 
 - Great way to visualize and identify statistically significant gene expression changes among hundreds to thousands of genes
   from different treatment conditions
   
**<span style="color:#33a8ff">How to create a heatmap using Python?</span>**

For generating a heatmap plot, I have used gene expression data published in Bedre et al. 2015 to identify 
gene expression changes (induced or downregulated) in response to fungal stress in cotton. 
(<a href="https://journals.plos.org/plosone/article?id=10.1371/journal.pone.0138025">Read paper</a>). 
Here's you can download gene expression dataset used for plotting heatmap: 
[dataset]({{"/myfiles/heatmap/hm_cot.csv" target="_blank""| absolute_url }})
 
 ```python
# We will use `bioinfokit` for plotting heatmap.
# download and install bioinfokit (Tested on Linux, Mac, Windows) 
# read instructions https://github.com/reneshbedre/bioinfokit
git clone https://github.com/reneshbedre/bioinfokit.git
cd bioinfokit
python setup.py install
```

Now plot heatmap with hierarchical clustering using bioinfokit,

```python
# you can use interactive python console, jupyter or python code
# I am using interactive python console (Python 3.6)
>>>from bioinfokit import visuz
# here you can change default parameters. 
# Read documentation at https://github.com/reneshbedre/bioinfokit
# heatmap with hierarchical clustering 
>>>visuz.hmap(table="hm_cot.csv", dim=(3, 6), tickfont=(6, 4))
# it needs csv file. If you have tab delimited file, you can reformat to
# csv using bioinfokit.analys.tcsv(file) 
# read docs at https://github.com/reneshbedre/bioinfokit

# heatmap without hierarchical clustering 
>>>visuz.hmap(table="hm_cot.csv", clus=False, dim=(3, 6), tickfont=(6, 4))
# heatmaps will be saved in same directory
```

Generated heatmaps with and without hierarchical clustering by above code,

<p align="center">
<img src="/myfiles/heatmap/heatmap_clus1.png" width="350">
<img src="/myfiles/heatmap/heatmap1.png" width="350">
</p>


<!-- [screenshot]({{ "/myfiles/heatmap/heatmap_clus.png" | absolute_url }})-->

The X-axis represents the treatment conditions and Y-axis represents the gene names. I have changed the names of 
six treatment conditions to A to F for the simplicity of understanding. You can <a href="https://journals.plos.org/plosone/article?id=10.1371/journal.pone.0138025">Read paper</a>
for a detailed understanding of the dataset. 

Now plot heatmap with different colormaps,
```python
# colormaps are available at  https://matplotlib.org/3.1.0/tutorials/colors/colormaps.html
# default is seismic 
# here I use red-yellow-green: RdYlGn
>>>visuz.hmap(table="hm_cot.csv", cmap='RdYlGn', dim=(3, 6), tickfont=(6, 4))
# it needs csv file. If you have tab delimited file, you can reformat to
# csv using bioinfokit.analys.tcsv(file) 
# read docs at https://github.com/reneshbedre/bioinfokit

# heatmap without hierarchical clustering 
>>>visuz.hmap(table="hm_cot.csv", clus=False, cmap='RdYlGn', dim=(3, 6), tickfont=(6, 4))
```

Generated heatmaps with a red-yellow-green colormap,

<p align="center">
<img src="/myfiles/heatmap/heatmap_clus2.png" width="350">
<img src="/myfiles/heatmap/heatmap2.png" width="350">
</p>

Now plot heatmap with standardized values,

```python
# Z-score can be used to standardize value with mean 0 and var 1
# default Z-score is set to None and it applies to only heatmap with cluster
# here I standardize column with Z-score
>>>visuz.hmap(table="hm_cot.csv", zscore=1, dim=(3, 6), tickfont=(6, 4))

# here I standardize row with Z-score
>>>visuz.hmap(table="hm_cot.csv", zscore=0, dim=(3, 6), tickfont=(6, 4))

```

Generated heatmaps with Z standardized column and row,

<p align="center">
<img src="/myfiles/heatmap/heatmap_clus31.png" width="350">
<img src="/myfiles/heatmap/heatmap_clus32.png" width="350">
</p>

In addition to these features, we can also control the label fontsize, 
figuresize, and scale of the heatmaps. Read full documentation at  
https://github.com/reneshbedre/bioinfokit


**<span style="color:#33a8ff">How to cite?</span>**

Bedre, R. Bioinformatics data analysis and visualization toolkit. 2018, GitHub repository, <a href="https://github.com/reneshbedre/bioinfokit">https://github.com/reneshbedre/bioinfokit</a>

<span style="color:#9e9696">If you have any questions, comments or recommendations, please email me at 
<b>reneshbe@gmail.com</b></span>
    
<span style="color:#9e9696"><i> Last updated: November 26, 2019</i> </span>    


<p>
{% include  share.html %}
</p>