---
title: "Merge multiple VCF files"
date:   2019-03-18 010:19:20
permalink: blog/mergevcf.html
author_profile: false
toc: true
toc_label: "Page Content"
---

<p>
{% include  share.html %}
</p>

## <span style="color:#33a8ff">What is VCF file?</span>
- VCF stands for variant call format
- It is a text file (file extension as .vcf) storing meta-information, marker and genotype data of genetic variations

## <span style="color:#33a8ff">How to merge multiple VCF files? </span>
- We will use `bioinfokit` 
- Check [bioinfokit documentation]({{"/blog/howtoinstall.html" | absolute_url }}) for installation and documentation

Sometimes, it is necessary to merge different VCF files for analysis as the genotype information stored in multiple 
files (For example, you have different VCF files for every chromosome).

```python
# you can use interactive python interpreter, jupyter notebook, spyder or python code
# I am using interactive python interpreter (Python 3.7)
# go to a directory where all vcf files are stored. make sure all files are uncompressed.
# make sure you will have uniform VCF files. For example, multiple VCF files 
# generated from same source datasets
>>>from bioinfokit import analys
# merge VCF files. You can provide multiple VCF files separated by comma.
>>>analys.mergevcf("file_1.vcf,file_2.vcf,file_3.vcf,file_4.vcf")
# merged VCF files will be stored in same directory (merge_vcf.vcf)
```


<p>
{% include  cite.html %}
</p>

<p>
{% include  share.html %}
</p>

<span style="color:#9e9696"><i> Last updated: March 18, 2019</i> </span>
