---
title: "Manhattan plot"
date:   2020-02-17 02:15:18
permalink: blog/manhat.html
author_profile: true
classes: wide
tags:
  - Genomic visualization
  - Variant analysis
  - High-throughput sequence analysis
---

<p>
{% include  share.html %}
</p>

**<span style="color:#33a8ff">What is Manhattan plot?</span>**
 - Manhattan plot used to visualize the association of SNPs with given trait or disease as statistical significance in 
   terms of P-values on a genomic scale 
 - In the Manhattan plot, X- and Y-axis represents the SNPs on the chromosomes and associated P-values as −log10[P].
 - It is a good way to visualize thousands to millions of SNPs o genome-scale. The lowest the P-value (higher −log10[P]), 
   highest is the association of a given SNP with trait or disease


**<span style="color:#33a8ff">How to create Manhattan plot?</span>**
- We will use `bioinfokit v0.9.2` or later
- Check [bioinfokit documentation]({{"/blog/howtoinstall.html" | absolute_url }}) for installation and documentation
- For generating Manhattan plot, I have used simulated GWAS data for 20K SNPs distributed over 10 chromosomes.
  Here's you can download GWAS dataset used for generating Manhattan plot: [dataset]({{"/assets/posts/mhat/gwas_res_sim.csv" | absolute_url }})

Useful reading: [Data handling using pandas ]({{"/blog/pandas.html" | absolute_url }})

```python
# you can use interactive python interpreter, jupyter notebook, spyder or python code
# I am using interactive python interpreter (Python 3.7)
>>> from bioinfokit import analys, visuz
# load dataset as pandas dataframe
>>> df = analys.get_data('mhat').data
>>> df.head()
   SNP    pvalue  chr
0  rs0  0.773739    3
1  rs1  0.554637    6
2  rs2  0.733446   10
3  rs3  0.872185    8
4  rs4  0.919034   10

# create Manhattan plot with default parameters
>>> visuz.marker.mhat(df=df, chr='chr',pv='pvalue')
# set parameter show=True, if you want view the image instead of saving
```

Generated Manhattan plot,

<p align="center">
<img src="/assets/posts/mhat/manhatten1.png" width="600">
</p>

Change colors
```python
# add alternate two colors
>>> visuz.marker.mhat(df=df, chr='chr',pv='pvalue', color=("#d7d1c9", "#696464"))
```

<p align="center">
<img src="/assets/posts/mhat/manhatten2.png" width="600">
</p>

```python
# add different colors equal to number of chromosomes
color=("#a7414a", "#696464", "#00743f", "#563838", "#6a8a82", "#a37c27", "#5edfff", "#282726", "#c0334d", "#c9753d")
>>> visuz.marker.mhat(df=df, chr='chr',pv='pvalue', color=color)
```

<p align="center">
<img src="/assets/posts/mhat/manhatten3.png" width="600">
</p>

Add genome-wide significance line,
```python
# by default line will be plotted at P=5E-08
# you can change this value as per need
>>> visuz.marker.mhat(df=df, chr='chr',pv='pvalue', color=color, gwas_sign_line=True)
```

<p align="center">
<img src="/assets/posts/mhat/manhatten4.png" width="600">
</p>

```python
# Change the position of genome-wide significance line
# you can change this value as per need
>>> visuz.marker.mhat(df=df, chr='chr',pv='pvalue', color=color, gwas_sign_line=True, gwasp=5E-06)
```
<p align="center">
<img src="/assets/posts/mhat/manhatten5.png" width="600">
</p>

Add annotation to SNPs (default text),

```python
# add name to SNPs based on the significance defined by 'gwasp'
>>> visuz.marker.mhat(df=df, chr='chr',pv='pvalue', color=color, gwas_sign_line=True, gwasp=5E-06, 
    markernames=True, markeridcol='SNP')
```

<p align="center">
<img src="/assets/posts/mhat/manhatten6.png" width="600">
</p>

Add annotation to SNPs (box text),

```python
# add name to SNPs based on the significance defined by 'gwasp'
>>> visuz.marker.mhat(df=df, chr='chr',pv='pvalue', color=color, gwas_sign_line=True, gwasp=5E-06, 
    markernames=True, markeridcol='SNP', gstyle=2)
```

<p align="center">
<img src="/assets/posts/mhat/manhatten_gs2_1.svg" width="600">
</p>

```python
# add name to specified  SNPs only
>>> visuz.marker.mhat(df=df, chr='chr',pv='pvalue', color=color, gwas_sign_line=True, gwasp=5E-06, 
    markernames=("rs19990", "rs40"), markeridcol='SNP')
```

<p align="center">
<img src="/assets/posts/mhat/manhatten7.png" width="600">
</p>

```python
# add name to specified  SNPs only (box text)
>>> visuz.marker.mhat(df=df, chr='chr',pv='pvalue', color=color, gwas_sign_line=True, gwasp=5E-06, 
    markernames=("rs19990", "rs40"), markeridcol='SNP', gstyle=2)
```

<p align="center">
<img src="/assets/posts/mhat/manhatten_gs2_2.svg" width="600">
</p>

```python
# change fontsize of SNP annotation
>>> visuz.marker.mhat(df=df, chr='chr',pv='pvalue', color=color, gwas_sign_line=True, gwasp=5E-06, markernames=True, 
    markeridcol='SNP', gfont=5)
# gfont is incompatible with gstyle    
```

<p align="center">
<img src="/assets/posts/mhat/manhatten13.png" width="600">
</p>

```python
# add gene names to SNPs
>>> visuz.marker.mhat(df=df, chr='chr',pv='pvalue', color=color, 
gwas_sign_line=True, gwasp=5E-06, markernames=({"rs19990":"gene1", "rs40":"gene2"}), markeridcol='SNP')
```

<p align="center">
<img src="/assets/posts/mhat/manhatten8.png" width="600">
</p>

Change fontsize, figure size, resolution, point size, and transparency

```python
# change figure size
>>> visuz.marker.mhat(df=df, chr='chr',pv='pvalue', color=color, dim=(8,6) )
```
<p align="center">
<img src="/assets/posts/mhat/manhatten9.png" width="600">
</p>

```python
# change point size
>>> visuz.marker.mhat(df=df, chr='chr',pv='pvalue', color=color, dotsize=2 )
```
<p align="center">
<img src="/assets/posts/mhat/manhatten10.png" width="600">
</p>

```python
# change point transparency
>>> visuz.marker.mhat(df=df, chr='chr',pv='pvalue', color=color, valpha=0.2 )
```
<p align="center">
<img src="/assets/posts/mhat/manhatten11.png" width="600">
</p>

```python
# change X-axis tick label rotation
>>> visuz.marker.mhat(df=df, chr='chr',pv='pvalue', color=color, ar=60 )
```
<p align="center">
<img src="/assets/posts/mhat/manhatten12.png" width="600">
</p>

```python
# change figure resolution
>>> visuz.marker.mhat(df=df, chr='chr',pv='pvalue', color=color, r=600 )
```
<p align="center">
<img src="/assets/posts/mhat/manhatten14.png" width="600">
</p>

In addition to these parameters, the parameters for figure type (`figtype`), Y axis ticks range (`ylm`), axis labels (`axxlabel`, `axylabel`),  
axis labels font size  (`axlabelfontsize``) can be provided.

Check detailed <a href='https://reneshbedre.github.io/blog/howtoinstall.html#manhatten-plot' target='_blank'>usage</a>


<p>
{% include  cite.html %}
</p>

<p>
{% include  share.html %}
</p>

<span style="color:#9e9696"><i> Last updated: July 30, 2020</i> </span>

<p>
{% include  license.html %}
</p>