<!doctype html>
<!--
  Minimal Mistakes Jekyll Theme 4.19.0 by Michael Rose
  Copyright 2013-2019 Michael Rose - mademistakes.com | @mmistakes
  Free for personal and commercial use under the MIT license
  https://github.com/mmistakes/minimal-mistakes/blob/master/LICENSE
-->
<html lang="en" class="no-js">
  <head>
    <meta charset="utf-8">

<!-- begin _includes/seo.html --><title>ANOVA using Python - Renesh Bedre</title>
<meta name="description" content="                             ">


  <meta name="author" content="Renesh Bedre">


<meta property="og:type" content="article">
<meta property="og:locale" content="en_US">
<meta property="og:site_name" content="Renesh Bedre">
<meta property="og:title" content="ANOVA using Python">
<meta property="og:url" content="https://reneshbedre.github.io//blog/anova.html">


  <meta property="og:description" content="                             ">





  <meta name="twitter:site" content="@renbedre">
  <meta name="twitter:title" content="ANOVA using Python">
  <meta name="twitter:description" content="                             ">
  <meta name="twitter:url" content="https://reneshbedre.github.io//blog/anova.html">

  
    <meta name="twitter:card" content="summary">
    
  

  



  <meta property="article:published_time" content="2018-10-22T01:18:08-05:00">





  

  


<link rel="canonical" href="https://reneshbedre.github.io//blog/anova.html">




<script type="application/ld+json">
  {
    "@context": "https://schema.org",
    
      "@type": "Person",
      "name": "Renesh Bedre",
      "url": "https://reneshbedre.github.io//"
    
  }
</script>






<!-- end _includes/seo.html -->


<link href="/feed.xml" type="application/atom+xml" rel="alternate" title="Renesh Bedre Feed">

<!-- https://t.co/dKP3o1e -->
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<script>
  document.documentElement.className = document.documentElement.className.replace(/\bno-js\b/g, '') + ' js ';
</script>

<!-- For all browsers -->
<link rel="stylesheet" href="/assets/css/main.css">

<!--[if IE]>
  <style>
    /* old IE unsupported flexbox fixes */
    .greedy-nav .site-title {
      padding-right: 3em;
    }
    .greedy-nav button {
      position: absolute;
      top: 0;
      right: 0;
      height: 100%;
    }
  </style>
<![endif]-->



    <!-- start custom head snippets -->

<!-- insert favicons. use https://realfavicongenerator.net/ -->

<!-- end custom head snippets -->

  </head>

  <body class="layout--single wide">
    <nav class="skip-links">
  <h2 class="screen-reader-text">Skip links</h2>
  <ul>
    <li><a href="#site-nav" class="screen-reader-shortcut">Skip to primary navigation</a></li>
    <li><a href="#main" class="screen-reader-shortcut">Skip to content</a></li>
    <li><a href="#footer" class="screen-reader-shortcut">Skip to footer</a></li>
  </ul>
</nav>

    <!--[if lt IE 9]>
<div class="notice--danger align-center" style="margin: 0;">You are using an <strong>outdated</strong> browser. Please <a href="https://browsehappy.com/">upgrade your browser</a> to improve your experience.</div>
<![endif]-->

    

<div class="masthead">
  <div class="masthead__inner-wrap">
    <div class="masthead__menu">
      <nav id="site-nav" class="greedy-nav">
        
        <a class="site-title" href="/">
          Renesh Bedre
          
        </a>
        <ul class="visible-links"><li class="masthead__menu-item">
              <a href="/about/">About</a>
            </li><li class="masthead__menu-item">
              <a href="/blog/">Blog</a>
            </li><li class="masthead__menu-item">
              <a href="/pub/">Publications</a>
            </li><li class="masthead__menu-item">
              <a href="/news/">News</a>
            </li><li class="masthead__menu-item">
              <a href="/res/">Resources</a>
            </li></ul>
        
        <button class="greedy-nav__toggle hidden" type="button">
          <span class="visually-hidden">Toggle menu</span>
          <div class="navicon"></div>
        </button>
        <ul class="hidden-links hidden"></ul>
      </nav>
    </div>
  </div>
</div>


    <div class="initial-content">
      



<div id="main" role="main">
  
  <div class="sidebar sticky">
  


<div itemscope itemtype="https://schema.org/Person">

  
    <div class="author__avatar">
      
        <img src="/assets/images/me.png" alt="Renesh Bedre" itemprop="image">
      
    </div>
  

  <div class="author__content">
    
      <h3 class="author__name" itemprop="name">Renesh Bedre</h3>
    
    
      <div class="author__bio" itemprop="description">
        <p>Bioinformatics Scientist</p>

      </div>
    
  </div>

  <div class="author__urls-wrapper">
    <button class="btn btn--inverse">Follow</button>
    <ul class="author__urls social-icons">
      
        <li itemprop="homeLocation" itemscope itemtype="https://schema.org/Place">
          <i class="fas fa-fw fa-map-marker-alt" aria-hidden="true"></i> <span itemprop="name">Texas A&M University</span>
        </li>
      

      
        
          
        
          
            <li><a href="https://reneshbedre.github.io/" rel="nofollow noopener noreferrer"><i class="fas fa-fw fa-link" aria-hidden="true"></i> Website</a></li>
          
        
          
            <li><a href="https://twitter.com/renbedre" rel="nofollow noopener noreferrer"><i class="fab fa-fw fa-twitter-square" aria-hidden="true"></i> Twitter</a></li>
          
        
          
            <li><a href="https://www.linkedin.com/in/reneshbedre/" rel="nofollow noopener noreferrer"><i class="fab fa-linkedin-in" aria-hidden="true"></i> LinkedIn</a></li>
          
        
          
        
          
            <li><a href="https://github.com/reneshbedre" rel="nofollow noopener noreferrer"><i class="fab fa-fw fa-github" aria-hidden="true"></i> GitHub</a></li>
          
        
          
        
          
            <li><a href="assets/scipapers/Bedre_CV.pdf" rel="nofollow noopener noreferrer"><i class="ai ai-cv-square" aria-hidden="true"></i> CV</a></li>
          
        
      

      

      
        <li>
          <a href="mailto:reneshbe@gmail.com">
            <meta itemprop="email" content="reneshbe@gmail.com" />
            <i class="fas fa-fw fa-envelope-square" aria-hidden="true"></i> Email
          </a>
        </li>
      

      

      

      

      

      

      

      

      

      

      

      

      

      

      

      

      

      

      

      

      

      

      

      <!--
  <li>
    <a href="http://link-to-whatever-social-network.com/user/" itemprop="sameAs" rel="nofollow noopener noreferrer">
      <i class="fas fa-fw" aria-hidden="true"></i> Custom Social Profile Link
    </a>
  </li>
-->
    </ul>
  </div>
</div>

  
  </div>



  <article class="page" itemscope itemtype="https://schema.org/CreativeWork">
    <meta itemprop="headline" content="ANOVA using Python">
    <meta itemprop="description" content="    ">
    <meta itemprop="datePublished" content="2018-10-22T01:18:08-05:00">
    

    <div class="page__inner-wrap">
      
        <header>
          <h1 id="page-title" class="page__title" itemprop="headline">ANOVA using Python
</h1>

          <!-- change by me to add date and read time together -->
          <p class="page__meta">
          
           <!-- <p class="page__meta"><i class="far fa-clock" aria-hidden="true"></i> 




  6 minute read

</p>-->

            
              <i class="fa fa-fw fa-calendar" aria-hidden="true"></i> <time datetime="2018-10-22T01:18:08-05:00">October 22, 2018 </time>&emsp;
          
          &nbsp;<i class="far fa-clock" aria-hidden="true"></i>&nbsp;




  6 minute read


          

          </p>
        </header>
      

      <section class="page__content" itemprop="text">
        
        <p>
<link rel="stylesheet" href="https://maxcdn.bootstrapcdn.com/font-awesome/4.5.0/css/font-awesome.min.css" />

<style>
    .share-box a {
  display: inline-block;
  -webkit-box-shadow: 0 0 1px #777;
  box-shadow: 0 0 1px #777;
  padding: 5px 12px;
  margin-right: 5px;
  margin-bottom: 5px;
  text-decoration: none; }
  .share-box a:hover {
    text-decoration: none;
    -webkit-transition: background-color 200ms linear;
    -ms-transition: background-color 200ms linear;
    transition: background-color 200ms linear; }

.f {
  color: #3b5998; }
  .f:hover {
    color: #fff;
    background-color: #3b5998; }

.t {
  color: #4099FF; }
  .t:hover {
    color: #fff;
    background-color: #4099FF; }

.g {
  color: #d34836; }
  .g:hover {
    color: #fff;
    background-color: #d34836; }

.r {
  color: #ff5700; }
  .r:hover {
    color: #fff;
    background-color: #ff5700; }

.l {
  color: #0077b5; }
  .l:hover {
    color: #fff;
    background-color: #0077b5; }

.e {
  color: #444444; }
  .e:hover {
    color: #fff;
    background-color: #444444; }
</style>

<div class="share-box">
<!--<h5>Share this:</h5>-->
<a class="f" href="https://www.facebook.com/sharer/sharer.php?u=https://reneshbedre.github.io//blog/anova.html" onclick="window.open(this.href, 'mywin',
'left=20,top=20,width=500,height=500,toolbar=1,resizable=0'); return false;"><i class="fab fa-facebook-f"></i></a>

<a class="t" href="https://twitter.com/intent/tweet?text=&amp;url=https://reneshbedre.github.io//blog/anova.html" onclick="window.open(this.href, 'mywin',
'left=20,top=20,width=500,height=500,toolbar=1,resizable=0'); return false;"><i class="fab fa-twitter"></i></a>

  <a class="g" href="https://plus.google.com/share?url=https://reneshbedre.github.io//blog/anova.html" onclick="window.open(this.href, 'mywin',
'left=20,top=20,width=500,height=500,toolbar=1,resizable=0'); return false;"><i class="fab fa-google-plus-square"></i></a>


<a class="r" href="http://www.reddit.com/submit?url=https://reneshbedre.github.io//blog/anova.html" onclick="window.open(this.href, 'mywin',
'left=20,top=20,width=900,height=500,toolbar=1,resizable=0'); return false;"><i class="fab fa-reddit"></i></a>

<a class="l" href="https://www.linkedin.com/shareArticle?mini=true&amp;url=https://reneshbedre.github.io//blog/anova.html" onclick="window.open(this.href, 'mywin',
'left=20,top=20,width=500,height=500,toolbar=1,resizable=0'); return false;"><i class="fab fa-linkedin-in"></i></a>

<a class="e" href="mailto:?subject=&amp;body=Check out this site https://reneshbedre.github.io//blog/anova.html"><i class="fas fa-envelope-square"></i></a>

  <!--
    <a href="https://twitter.com/renbedre?ref_src=twsrc%5Etfw" class="twitter-follow-button" data-show-count="false">Follow @renbedre</a><script async src="https://platform.twitter.com/widgets.js" charset="utf-8"></script>
-->

</div>
</p>

<p><strong><span style="color:#33a8ff">What is ANOVA (ANalysis Of VAriance)?</span></strong></p>
<ul>
  <li>used to compare the means of more than 2 groups (t-test can be used to compare 2 groups)</li>
  <li>groups mean differences inferred by analyzing variances</li>
  <li>Main types: One-way (one factor) and two-way (two factors) ANOVA (factor is an independent variable)</li>
</ul>

<p>Note: In ANOVA, group, factors, and independent variables are similar terms</p>

<p><strong><span style="color:#33a8ff">ANOVA Hypotheses</span></strong></p>
<ul>
  <li><i>Null hypotheses</i>: Groups means are equal (no variation in means of groups)</li>
  <li><i>Alternative hypotheses</i>: At least, one group mean is different from other groups</li>
</ul>

<p><strong><span style="color:#33a8ff">ANOVA Assumptions</span></strong></p>
<ul>
  <li>Residuals (experimental error) are normally distributed (Shapiro Wilks Test)</li>
  <li>Homogeneity of variances (variances are equal between treatment groups) (Levene or Bartlett Test)</li>
  <li>Observations are sampled independently from each other</li>
</ul>

<p><strong><span style="color:#33a8ff">How ANOVA works?</span></strong></p>
<ul>
  <li>Check sample sizes: equal number of observation in each group</li>
  <li>Calculate Mean Square for each group (MS) (SS of group/level-1);
level-1 is a degree of freedom (df) for a group</li>
  <li>Calculate Mean Square error (MSE) (SS error/df of residuals)</li>
  <li>Calculate F-value (MS of group/MSE)</li>
</ul>

<p><strong><span style="color:#33a8ff">One-way (one factor) ANOVA</span></strong></p>

<p>Example data for one-way ANOVA analysis, <a href="https://reneshbedre.github.io//assets/posts/anova/onewayanova.txt">dataset</a></p>

<table>
  <thead>
    <tr>
      <th>A</th>
      <th>B</th>
      <th>C</th>
      <th>D</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td>25</td>
      <td>45</td>
      <td>30</td>
      <td>54</td>
    </tr>
    <tr>
      <td>30</td>
      <td>55</td>
      <td>29</td>
      <td>60</td>
    </tr>
    <tr>
      <td>28</td>
      <td>29</td>
      <td>33</td>
      <td>51</td>
    </tr>
    <tr>
      <td>36</td>
      <td>56</td>
      <td>37</td>
      <td>62</td>
    </tr>
    <tr>
      <td>29</td>
      <td>40</td>
      <td>27</td>
      <td>73</td>
    </tr>
  </tbody>
</table>

<p>Here, there are four treatments (A, B, C, and D), which are groups for ANOVA analysis. Treatments are independent 
variable and termed as factor. As there are four types of treatments, treatment factor has four levels.</p>

<p>For this experimental design, there is only factor (treatments) or independent variable to evaluate, and therefore, 
one-way ANOVA is suitable for analysis.</p>

<p>Useful reading: <a href="https://reneshbedre.github.io//blog/pandas.html">Data handling using pandas </a></p>

<div class="language-python highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="c1"># load packages
</span><span class="kn">import</span> <span class="nn">pandas</span> <span class="k">as</span> <span class="n">pd</span>
<span class="c1"># load data file
</span><span class="n">d</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="s">"https://reneshbedre.github.io/assets/posts/anova/onewayanova.txt"</span><span class="p">,</span> <span class="n">sep</span><span class="o">=</span><span class="s">"</span><span class="se">\t</span><span class="s">"</span><span class="p">)</span>
<span class="c1"># generate a boxplot to see the data distribution by treatments. Using boxplot, we can easily detect the differences 
# between different treatments
</span><span class="n">d</span><span class="o">.</span><span class="n">boxplot</span><span class="p">(</span><span class="n">column</span><span class="o">=</span><span class="p">[</span><span class="s">'A'</span><span class="p">,</span> <span class="s">'B'</span><span class="p">,</span> <span class="s">'C'</span><span class="p">,</span> <span class="s">'D'</span><span class="p">],</span> <span class="n">grid</span><span class="o">=</span><span class="bp">False</span><span class="p">)</span>
</code></pre></div></div>

<p><img src="https://reneshbedre.github.io//assets/posts/anova/onewaynova_boxplot.png" alt="screenshot" /></p>

<div class="language-python highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="c1"># load packages
</span><span class="kn">import</span> <span class="nn">scipy.stats</span> <span class="k">as</span> <span class="n">stats</span>
<span class="c1"># stats f_oneway functions takes the groups as input and returns F and P-value
</span><span class="n">fvalue</span><span class="p">,</span> <span class="n">pvalue</span> <span class="o">=</span> <span class="n">stats</span><span class="o">.</span><span class="n">f_oneway</span><span class="p">(</span><span class="n">d</span><span class="p">[</span><span class="s">'A'</span><span class="p">],</span> <span class="n">d</span><span class="p">[</span><span class="s">'B'</span><span class="p">],</span> <span class="n">d</span><span class="p">[</span><span class="s">'C'</span><span class="p">],</span> <span class="n">d</span><span class="p">[</span><span class="s">'D'</span><span class="p">])</span>
<span class="k">print</span><span class="p">(</span><span class="n">fvalue</span><span class="p">,</span> <span class="n">pvalue</span><span class="p">)</span>
<span class="c1"># 17.492810457516338 2.639241146210922e-05
</span>
<span class="c1"># get ANOVA table as R like output
</span><span class="kn">import</span> <span class="nn">statsmodels.api</span> <span class="k">as</span> <span class="n">sm</span>
<span class="kn">from</span> <span class="nn">statsmodels.formula.api</span> <span class="kn">import</span> <span class="n">ols</span>
<span class="c1"># reshape the d dataframe suitable for statsmodels package 
</span><span class="n">d_melt</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">melt</span><span class="p">(</span><span class="n">d</span><span class="o">.</span><span class="n">reset_index</span><span class="p">(),</span> <span class="n">id_vars</span><span class="o">=</span><span class="p">[</span><span class="s">'index'</span><span class="p">],</span> <span class="n">value_vars</span><span class="o">=</span><span class="p">[</span><span class="s">'A'</span><span class="p">,</span> <span class="s">'B'</span><span class="p">,</span> <span class="s">'C'</span><span class="p">,</span> <span class="s">'D'</span><span class="p">])</span>
<span class="c1"># replace column names
</span><span class="n">d_melt</span><span class="o">.</span><span class="n">columns</span> <span class="o">=</span> <span class="p">[</span><span class="s">'index'</span><span class="p">,</span> <span class="s">'treatments'</span><span class="p">,</span> <span class="s">'value'</span><span class="p">]</span>
<span class="c1"># Ordinary Least Squares (OLS) model
</span><span class="n">model</span> <span class="o">=</span> <span class="n">ols</span><span class="p">(</span><span class="s">'value ~ C(treatments)'</span><span class="p">,</span> <span class="n">data</span><span class="o">=</span><span class="n">d_melt</span><span class="p">)</span><span class="o">.</span><span class="n">fit</span><span class="p">()</span>
<span class="n">anova_table</span> <span class="o">=</span> <span class="n">sm</span><span class="o">.</span><span class="n">stats</span><span class="o">.</span><span class="n">anova_lm</span><span class="p">(</span><span class="n">model</span><span class="p">,</span> <span class="n">typ</span><span class="o">=</span><span class="mi">2</span><span class="p">)</span>
<span class="n">anova_table</span>

<span class="o">|</span> <span class="n">df</span>            <span class="o">|</span> <span class="n">sum_sq</span> <span class="o">|</span> <span class="n">mean_sq</span> <span class="o">|</span> <span class="n">F</span>        <span class="o">|</span> <span class="n">PR</span><span class="p">(</span><span class="o">&gt;</span><span class="n">F</span><span class="p">)</span>   <span class="o">|</span>          <span class="o">|</span>
<span class="o">|---------------|--------|---------|----------|----------|----------|</span>
<span class="o">|</span> <span class="n">C</span><span class="p">(</span><span class="n">treatments</span><span class="p">)</span> <span class="o">|</span> <span class="mf">3.0</span>    <span class="o">|</span> <span class="mf">3010.95</span> <span class="o">|</span> <span class="mf">1003.650</span> <span class="o">|</span> <span class="mf">17.49281</span> <span class="o">|</span> <span class="mf">0.000026</span> <span class="o">|</span>
<span class="o">|</span> <span class="n">Residual</span>      <span class="o">|</span> <span class="mf">16.0</span>   <span class="o">|</span> <span class="mf">918.00</span>  <span class="o">|</span> <span class="mf">57.375</span>   <span class="o">|</span> <span class="n">NaN</span>      <span class="o">|</span> <span class="n">NaN</span>      <span class="o">|</span>
</code></pre></div></div>

<p><b>Interpretation</b>: The P-value obtained from ANOVA analysis is significant (P&lt;0.05), and therefore, we conclude 
that there are significant differences among treatments.</p>

<p>From ANOVA analysis, we know that treatment differences are statistically significant, but ANOVA does not tell which 
treatments are significantly different from each other. To know the pairs of significant different treatments, we
will perform multiple pairwise comparison (<b>Post-hoc comparison</b>) analysis using <b>Tukey HSD</b> test.</p>

<div class="language-python highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="c1"># load packages
</span><span class="kn">from</span> <span class="nn">statsmodels.stats.multicomp</span> <span class="kn">import</span> <span class="n">pairwise_tukeyhsd</span>
<span class="c1"># perform multiple pairwise comparison (Tukey HSD)
</span><span class="n">m_comp</span> <span class="o">=</span> <span class="n">pairwise_tukeyhsd</span><span class="p">(</span><span class="n">endog</span><span class="o">=</span><span class="n">d_melt</span><span class="p">[</span><span class="s">'value'</span><span class="p">],</span> <span class="n">groups</span><span class="o">=</span><span class="n">d_melt</span><span class="p">[</span><span class="s">'treatments'</span><span class="p">],</span> <span class="n">alpha</span><span class="o">=</span><span class="mf">0.05</span><span class="p">)</span>
<span class="k">print</span><span class="p">(</span><span class="n">m_comp</span><span class="p">)</span>

<span class="n">Multiple</span> <span class="n">Comparison</span> <span class="n">of</span> <span class="n">Means</span> <span class="o">-</span> <span class="n">Tukey</span> <span class="n">HSD</span><span class="p">,</span><span class="n">FWER</span><span class="o">=</span><span class="mf">0.05</span>
<span class="o">==============================================</span>
<span class="n">group1</span> <span class="n">group2</span> <span class="n">meandiff</span>  <span class="n">lower</span>    <span class="n">upper</span>  <span class="n">reject</span>
<span class="o">----------------------------------------------</span>
  <span class="n">A</span>      <span class="n">B</span>      <span class="mf">15.4</span>    <span class="mf">1.6929</span>  <span class="mf">29.1071</span>  <span class="bp">True</span> 
  <span class="n">A</span>      <span class="n">C</span>      <span class="mf">1.6</span>    <span class="o">-</span><span class="mf">12.1071</span> <span class="mf">15.3071</span> <span class="bp">False</span> 
  <span class="n">A</span>      <span class="n">D</span>      <span class="mf">30.4</span>   <span class="mf">16.6929</span>  <span class="mf">44.1071</span>  <span class="bp">True</span> 
  <span class="n">B</span>      <span class="n">C</span>     <span class="o">-</span><span class="mf">13.8</span>   <span class="o">-</span><span class="mf">27.5071</span> <span class="o">-</span><span class="mf">0.0929</span>  <span class="bp">True</span> 
  <span class="n">B</span>      <span class="n">D</span>      <span class="mf">15.0</span>    <span class="mf">1.2929</span>  <span class="mf">28.7071</span>  <span class="bp">True</span> 
  <span class="n">C</span>      <span class="n">D</span>      <span class="mf">28.8</span>   <span class="mf">15.0929</span>  <span class="mf">42.5071</span>  <span class="bp">True</span> 
<span class="o">----------------------------------------------</span>
</code></pre></div></div>

<p>Above results from Tukey HSD suggests that except A-C, all other pairwise comparisons for treatments rejects null 
hypothesis and indicates statistical significant differences.</p>

<p><strong>Use bioinfokit for performing <a href="https://reneshbedre.github.io//blog/oanova.html">one-way ANOVA analysis</a> in Python</strong></p>

<p><strong><span style="color:#33a8ff">Test ANOVA assumptions</span></strong></p>

<p>The <b>Shapiro-Wilk test</b> can be used to check the <b> normal distribution of residuals </b>. <i>Null hypothesis</i>: 
data is drawn from normal distribution.</p>

<div class="language-python highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="c1"># load packages
</span><span class="kn">import</span> <span class="nn">scipy.stats</span> <span class="k">as</span> <span class="n">stats</span>
<span class="n">w</span><span class="p">,</span> <span class="n">pvalue</span> <span class="o">=</span> <span class="n">stats</span><span class="o">.</span><span class="n">shapiro</span><span class="p">(</span><span class="n">model</span><span class="o">.</span><span class="n">resid</span><span class="p">)</span>
<span class="k">print</span><span class="p">(</span><span class="n">w</span><span class="p">,</span> <span class="n">pvalue</span><span class="p">)</span>
<span class="c1"># 0.9685019850730896 0.7229772806167603
</span></code></pre></div></div>

<p>As the P-value is non significant, we fail to reject null hypothesis and
conclude that data is drawn from normal distribution.</p>

<p>As the data is drawn from normal distribution, use Bartlett’s test 
to check the <b>Homogeneity of variances</b>.
 <i>Null hypothesis</i>: samples from populations 
have equal variances.</p>

<div class="language-python highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="c1"># load packages
</span><span class="kn">import</span> <span class="nn">scipy.stats</span> <span class="k">as</span> <span class="n">stats</span>
<span class="n">w</span><span class="p">,</span> <span class="n">pvalue</span> <span class="o">=</span> <span class="n">stats</span><span class="o">.</span><span class="n">bartlett</span><span class="p">(</span><span class="n">d</span><span class="p">[</span><span class="s">'A'</span><span class="p">],</span> <span class="n">d</span><span class="p">[</span><span class="s">'B'</span><span class="p">],</span> <span class="n">d</span><span class="p">[</span><span class="s">'C'</span><span class="p">],</span> <span class="n">d</span><span class="p">[</span><span class="s">'D'</span><span class="p">])</span>
<span class="k">print</span><span class="p">(</span><span class="n">w</span><span class="p">,</span> <span class="n">pvalue</span><span class="p">)</span>
<span class="mf">5.687843565012841</span> <span class="mf">0.1278253399753447</span>
</code></pre></div></div>
<p>As the P-value (0.12) is non significant, we fail to reject null 
hypothesis and conclude that treatments have equal variances.</p>

<p><b>Levene test</b> can be used to check the Homogeneity of variances
when the data is not drawn from normal distribution.</p>

<!--
```python
# load packages
import scipy.stats as stats
w, pvalue = stats.levene(d['A'], d['B'], d['C'], d['D'])
print(w, pvalue)
# 1.9219593192195938 0.16673281219949276
```
-->

<p><strong><span style="color:#33a8ff">Two-way (two factor) ANOVA</span></strong></p>

<p>Example data for two-way ANOVA analysis, <a href="https://reneshbedre.github.io//assets/posts/anova/twowayanova.txt">dataset</a></p>

<p>From dataset, there are two factors (independent variables) viz. genotypes and yield in years. Genotypes and years has 
five and three levels respectively (see one-way ANOVA to know factors and levels).</p>

<p>For this experimental design, there are two factors to evaluate, and therefore, two-way ANOVA is suitable for analysis. 
Here, using two-way ANOVA, we can simultaneously evaluate how type of genotype and years affects the yields of plants. If
you apply one-way ANOVA here, you can able to evaluate only one factor at a time.</p>

<p>From two-way ANOVA, we can tests three hypotheses 1) effect of genotype on yield 2) effect of time (years) on yield, and
3) effect of genotype and time (years) interactions on yield</p>

<div class="language-python highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="c1"># load packages
</span><span class="kn">import</span> <span class="nn">pandas</span> <span class="k">as</span> <span class="n">pd</span>
<span class="kn">import</span> <span class="nn">seaborn</span> <span class="k">as</span> <span class="n">sns</span>
<span class="c1"># load data file
</span><span class="n">d</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="s">"https://reneshbedre.github.io/assets/posts/anova/twowayanova.txt"</span><span class="p">,</span> <span class="n">sep</span><span class="o">=</span><span class="s">"</span><span class="se">\t</span><span class="s">"</span><span class="p">)</span>
<span class="c1"># reshape the d dataframe suitable for statsmodels package 
# you do not need to reshape if your data is already in stacked format. Compare d and d_melt tables for detail 
# understanding 
</span><span class="n">d_melt</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">melt</span><span class="p">(</span><span class="n">d</span><span class="p">,</span> <span class="n">id_vars</span><span class="o">=</span><span class="p">[</span><span class="s">'Genotype'</span><span class="p">],</span> <span class="n">value_vars</span><span class="o">=</span><span class="p">[</span><span class="s">'1_year'</span><span class="p">,</span> <span class="s">'2_year'</span><span class="p">,</span> <span class="s">'3_year'</span><span class="p">])</span>
<span class="c1"># replace column names
</span><span class="n">d_melt</span><span class="o">.</span><span class="n">columns</span> <span class="o">=</span> <span class="p">[</span><span class="s">'Genotype'</span><span class="p">,</span> <span class="s">'years'</span><span class="p">,</span> <span class="s">'value'</span><span class="p">]</span>
<span class="c1"># generate a boxplot to see the data distribution by genotypes and years. Using boxplot, we can easily detect the 
# differences between different groups
</span><span class="n">sns</span><span class="o">.</span><span class="n">boxplot</span><span class="p">(</span><span class="n">x</span><span class="o">=</span><span class="s">"Genotype"</span><span class="p">,</span> <span class="n">y</span><span class="o">=</span><span class="s">"value"</span><span class="p">,</span> <span class="n">hue</span><span class="o">=</span><span class="s">"years"</span><span class="p">,</span> <span class="n">data</span><span class="o">=</span><span class="n">d_melt</span><span class="p">,</span> <span class="n">palette</span><span class="o">=</span><span class="s">"Set3"</span><span class="p">)</span> 
</code></pre></div></div>

<p><img src="https://reneshbedre.github.io//assets/posts/anova/twowayanova_boxplot.png" alt="screenshot" /></p>

<div class="language-python highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="c1"># load packages
</span><span class="kn">import</span> <span class="nn">statsmodels.api</span> <span class="k">as</span> <span class="n">sm</span>
<span class="kn">from</span> <span class="nn">statsmodels.formula.api</span> <span class="kn">import</span> <span class="n">ols</span>
<span class="c1"># Ordinary Least Squares (OLS) model
# C(Genotype):C(years) represent interaction term
</span><span class="n">model</span> <span class="o">=</span> <span class="n">ols</span><span class="p">(</span><span class="s">'value ~ C(Genotype) + C(years) + C(Genotype):C(years)'</span><span class="p">,</span> <span class="n">data</span><span class="o">=</span><span class="n">d_melt</span><span class="p">)</span><span class="o">.</span><span class="n">fit</span><span class="p">()</span>
<span class="n">anova_table</span> <span class="o">=</span> <span class="n">sm</span><span class="o">.</span><span class="n">stats</span><span class="o">.</span><span class="n">anova_lm</span><span class="p">(</span><span class="n">model</span><span class="p">,</span> <span class="n">typ</span><span class="o">=</span><span class="mi">2</span><span class="p">)</span>
<span class="n">anova_table</span>

<span class="o">|</span>                      <span class="o">|</span> <span class="n">sum_sq</span>     <span class="o">|</span> <span class="n">df</span>   <span class="o">|</span>     <span class="n">F</span>      <span class="o">|</span>    <span class="n">PR</span><span class="p">(</span><span class="o">&gt;</span><span class="n">F</span><span class="p">)</span>    <span class="o">|</span>
<span class="o">|----------------------|------------|------|------------|--------------|</span>
<span class="o">|</span> <span class="n">C</span><span class="p">(</span><span class="n">Genotype</span><span class="p">)</span>          <span class="o">|</span> <span class="mf">58.551733</span>  <span class="o">|</span> <span class="mf">5.0</span>  <span class="o">|</span> <span class="mf">32.748581</span>  <span class="o">|</span> <span class="mf">1.931655e-12</span> <span class="o">|</span>
<span class="o">|</span> <span class="n">C</span><span class="p">(</span><span class="n">years</span><span class="p">)</span>             <span class="o">|</span> <span class="mf">278.925633</span> <span class="o">|</span> <span class="mf">2.0</span>  <span class="o">|</span> <span class="mf">390.014868</span> <span class="o">|</span> <span class="mf">4.006243e-25</span> <span class="o">|</span>
<span class="o">|</span> <span class="n">C</span><span class="p">(</span><span class="n">Genotype</span><span class="p">):</span><span class="n">C</span><span class="p">(</span><span class="n">years</span><span class="p">)</span> <span class="o">|</span> <span class="mf">17.122967</span>  <span class="o">|</span> <span class="mf">10.0</span> <span class="o">|</span> <span class="mf">4.788525</span>   <span class="o">|</span> <span class="mf">2.230094e-04</span> <span class="o">|</span>
<span class="o">|</span> <span class="n">Residual</span>             <span class="o">|</span> <span class="mf">12.873000</span>  <span class="o">|</span> <span class="mf">36.0</span> <span class="o">|</span> <span class="n">NaN</span>        <span class="o">|</span> <span class="n">NaN</span>          <span class="o">|</span>
</code></pre></div></div>

<p><b>Interpretation</b>: The P-value obtained from ANOVA analysis for genotype, years, and interaction are statistically 
significant (P&lt;0.05). We conclude that type of genotype significantly affects the yield outcome, time (years) 
significantly affects the yield outcome, and interaction of both genotype and time (years) significantly affects the 
yield outcome.</p>

<p>Now, we know that genotype and time (years) differences are statistically significant, but ANOVA does not tell which 
genotype and time (years) are significantly different from each other. To know the pairs of significant different 
genotype and time (years), perform multiple pairwise comparison (<b>Post-hoc comparison</b>) analysis using 
<b>Tukey HSD</b> test.</p>

<p>Similar to one-way ANOVA, you can use <b>Levene</b> and <b>Shapiro-Wilk test</b> to validate the assumptions for
homogeneity of variances and normal distribution of residuals.</p>

<p><strong><span style="color:#33a8ff">Additional Notes</span></strong></p>
<ul>
  <li>three factor designs can be analyzed in a similar way to two-way ANOVA</li>
</ul>

<p><strong><span style="color:#33a8ff">How to cite?</span></strong></p>

<p>Bedre, R. Bioinformatics data analysis and visualization toolkit. 
GitHub repository, <a href="https://github.com/reneshbedre/bioinfokit">https://github.com/reneshbedre/bioinfokit</a></p>

<p>
<link rel="stylesheet" href="https://maxcdn.bootstrapcdn.com/font-awesome/4.5.0/css/font-awesome.min.css" />

<style>
    .share-box a {
  display: inline-block;
  -webkit-box-shadow: 0 0 1px #777;
  box-shadow: 0 0 1px #777;
  padding: 5px 12px;
  margin-right: 5px;
  margin-bottom: 5px;
  text-decoration: none; }
  .share-box a:hover {
    text-decoration: none;
    -webkit-transition: background-color 200ms linear;
    -ms-transition: background-color 200ms linear;
    transition: background-color 200ms linear; }

.f {
  color: #3b5998; }
  .f:hover {
    color: #fff;
    background-color: #3b5998; }

.t {
  color: #4099FF; }
  .t:hover {
    color: #fff;
    background-color: #4099FF; }

.g {
  color: #d34836; }
  .g:hover {
    color: #fff;
    background-color: #d34836; }

.r {
  color: #ff5700; }
  .r:hover {
    color: #fff;
    background-color: #ff5700; }

.l {
  color: #0077b5; }
  .l:hover {
    color: #fff;
    background-color: #0077b5; }

.e {
  color: #444444; }
  .e:hover {
    color: #fff;
    background-color: #444444; }
</style>

<div class="share-box">
<!--<h5>Share this:</h5>-->
<a class="f" href="https://www.facebook.com/sharer/sharer.php?u=https://reneshbedre.github.io//blog/anova.html" onclick="window.open(this.href, 'mywin',
'left=20,top=20,width=500,height=500,toolbar=1,resizable=0'); return false;"><i class="fab fa-facebook-f"></i></a>

<a class="t" href="https://twitter.com/intent/tweet?text=&amp;url=https://reneshbedre.github.io//blog/anova.html" onclick="window.open(this.href, 'mywin',
'left=20,top=20,width=500,height=500,toolbar=1,resizable=0'); return false;"><i class="fab fa-twitter"></i></a>

  <a class="g" href="https://plus.google.com/share?url=https://reneshbedre.github.io//blog/anova.html" onclick="window.open(this.href, 'mywin',
'left=20,top=20,width=500,height=500,toolbar=1,resizable=0'); return false;"><i class="fab fa-google-plus-square"></i></a>


<a class="r" href="http://www.reddit.com/submit?url=https://reneshbedre.github.io//blog/anova.html" onclick="window.open(this.href, 'mywin',
'left=20,top=20,width=900,height=500,toolbar=1,resizable=0'); return false;"><i class="fab fa-reddit"></i></a>

<a class="l" href="https://www.linkedin.com/shareArticle?mini=true&amp;url=https://reneshbedre.github.io//blog/anova.html" onclick="window.open(this.href, 'mywin',
'left=20,top=20,width=500,height=500,toolbar=1,resizable=0'); return false;"><i class="fab fa-linkedin-in"></i></a>

<a class="e" href="mailto:?subject=&amp;body=Check out this site https://reneshbedre.github.io//blog/anova.html"><i class="fas fa-envelope-square"></i></a>

  <!--
    <a href="https://twitter.com/renbedre?ref_src=twsrc%5Etfw" class="twitter-follow-button" data-show-count="false">Follow @renbedre</a><script async src="https://platform.twitter.com/widgets.js" charset="utf-8"></script>
-->

</div>
</p>

<p><span style="color:#9e9696">If you have any questions, comments or recommendations, please email me at 
<b>reneshbe@gmail.com</b></span></p>

<p><span style="color:#9e9696"><i> Last updated: January 31, 2020</i> </span></p>


        
      </section>




      <footer class="page__meta">
        
        


        <!-- commented to remove updated in footer section
        
          <p class="page__date"><strong><i class="fas fa-fw fa-calendar-alt" aria-hidden="true"></i> Updated:</strong> <time datetime="2018-10-22T01:18:08-05:00">October 22, 2018</time></p>
        
        -->
      </footer>

      <section class="page__share">
  
    <h4 class="page__share-title">Share on</h4>
  

  <a href="https://twitter.com/intent/tweet?via=renbedre&text=ANOVA+using+Python%20https%3A%2F%2Freneshbedre.github.io%2F%2Fblog%2Fanova.html" class="btn btn--twitter" onclick="window.open(this.href, 'window', 'left=20,top=20,width=500,height=500,toolbar=1,resizable=0'); return false;" title="Share on Twitter"><i class="fab fa-fw fa-twitter" aria-hidden="true"></i><span> Twitter</span></a>

  <a href="https://www.facebook.com/sharer/sharer.php?u=https%3A%2F%2Freneshbedre.github.io%2F%2Fblog%2Fanova.html" class="btn btn--facebook" onclick="window.open(this.href, 'window', 'left=20,top=20,width=500,height=500,toolbar=1,resizable=0'); return false;" title="Share on Facebook"><i class="fab fa-fw fa-facebook" aria-hidden="true"></i><span> Facebook</span></a>

  <a href="https://www.linkedin.com/shareArticle?mini=true&url=https%3A%2F%2Freneshbedre.github.io%2F%2Fblog%2Fanova.html" class="btn btn--linkedin" onclick="window.open(this.href, 'window', 'left=20,top=20,width=500,height=500,toolbar=1,resizable=0'); return false;" title="Share on LinkedIn"><i class="fab fa-fw fa-linkedin" aria-hidden="true"></i><span> LinkedIn</span></a>
</section>


      
  <nav class="pagination">
    
      <a href="/blog/pca_3d.html" class="pagination--pager" title="Principal component analysis (PCA) analysis and visualization using Python
">Previous</a>
    
    
      <a href="/blog/volcano.html" class="pagination--pager" title="Volcano plot in Python
">Next</a>
    
  </nav>

    </div>

    
  </article>

  
  
    <div class="page__related">
      <h4 class="page__related-title">You may also enjoy</h4>
      <div class="grid__wrapper">
        
          



<div class="grid__item">
  <article class="archive__item" itemscope itemtype="https://schema.org/CreativeWork">
    
    <h2 class="archive__item-title" itemprop="headline">
      
        <a href="/blog/howtoinstall.html" rel="permalink">How to install bioinfokit
</a>
      
    </h2>
    
      <p class="page__meta"><i class="far fa-clock" aria-hidden="true"></i> 




  less than 1 minute read

</p>
    
    <p class="archive__item-excerpt" itemprop="description">










  








  




</p>
  </article>
</div>

        
          



<div class="grid__item">
  <article class="archive__item" itemscope itemtype="https://schema.org/CreativeWork">
    
    <h2 class="archive__item-title" itemprop="headline">
      
        <a href="/blog/fqqualfmt.html" rel="permalink">FASTQ Sequence Quality Format
</a>
      
    </h2>
    
      <p class="page__meta"><i class="far fa-clock" aria-hidden="true"></i> 




  3 minute read

</p>
    
    <p class="archive__item-excerpt" itemprop="description">

</p>
  </article>
</div>

        
          



<div class="grid__item">
  <article class="archive__item" itemscope itemtype="https://schema.org/CreativeWork">
    
    <h2 class="archive__item-title" itemprop="headline">
      
        <a href="/blog/bardot.html" rel="permalink">Bar graph with overlapping dots (Bar-dot) plot
</a>
      
    </h2>
    
      <p class="page__meta"><i class="far fa-clock" aria-hidden="true"></i> 




  1 minute read

</p>
    
    <p class="archive__item-excerpt" itemprop="description">










  








  




</p>
  </article>
</div>

        
          



<div class="grid__item">
  <article class="archive__item" itemscope itemtype="https://schema.org/CreativeWork">
    
    <h2 class="archive__item-title" itemprop="headline">
      
        <a href="/blog/manhat.html" rel="permalink">Manhattan plot
</a>
      
    </h2>
    
      <p class="page__meta"><i class="far fa-clock" aria-hidden="true"></i> 




  2 minute read

</p>
    
    <p class="archive__item-excerpt" itemprop="description">










  








  




</p>
  </article>
</div>

        
      </div>
    </div>
  
  
</div>

    </div>

    

    <div id="footer" class="page__footer">
      <footer>
        <!-- start custom footer snippets -->

<!-- end custom footer snippets -->
        <link rel="stylesheet" href="https://cdn.rawgit.com/jpswalsh/academicons/master/css/academicons.min.css">

<div class="page__footer-follow">
  <ul class="social-icons">
    
      <li><strong>Follow:</strong></li>
    

    
      
        
          <li><a href="https://twitter.com/renbedre" rel="nofollow noopener noreferrer"><i class="fab fa-fw fa-twitter-square" aria-hidden="true"></i> Twitter</a></li>
        
      
        
      
        
          <li><a href="https://github.com/reneshbedre" rel="nofollow noopener noreferrer"><i class="fab fa-fw fa-github" aria-hidden="true"></i> GitHub</a></li>
        
      
        
      
        
      
        
      
        
          <li><a href="https://scholar.google.com/citations?user=6pv2AiMAAAAJ&hl=en" rel="nofollow noopener noreferrer"><i class="ai ai-google-scholar-square" aria-hidden="true"></i> Google Scholar</a></li>
        
      
        
          <li><a href="https://scholar.google.com/citations?user=6pv2AiMAAAAJ&hl=en" rel="nofollow noopener noreferrer"><i class="ai ai-researchgate-square" aria-hidden="true"></i> ResearchGate</a></li>
        
      
    

    <li><a href="/feed.xml"><i class="fas fa-fw fa-rss-square" aria-hidden="true"></i> Feed</a></li>
  </ul>
</div>

<div class="page__footer-copyright">&copy; 2020 Renesh Bedre. Powered by <a href="https://jekyllrb.com" rel="nofollow">Jekyll</a> &amp; <a href="https://mademistakes.com/work/minimal-mistakes-jekyll-theme/" rel="nofollow">Minimal Mistakes</a>.</div>

      </footer>
    </div>

    
  <script src="/assets/js/main.min.js"></script>
  <script src="https://kit.fontawesome.com/4eee35f757.js"></script>







  <!-- Global site tag (gtag.js) - Google Analytics -->
<script async src="https://www.googletagmanager.com/gtag/js?id=UA-115112853-1"></script>
<script>
  window.dataLayer = window.dataLayer || [];
  function gtag(){dataLayer.push(arguments);}
  gtag('js', new Date());

  gtag('config', 'UA-115112853-1', { 'anonymize_ip': false});
</script>


<!-- Begin Inspectlet Asynchronous Code -->
    <script type="text/javascript">
    (function() {
    window.__insp = window.__insp || [];
    __insp.push(['wid', 814107329]);
    var ldinsp = function(){
    if(typeof window.__inspld != "undefined") return; window.__inspld = 1; var insp = document.createElement('script'); insp.type = 'text/javascript'; insp.async = true; insp.id = "inspsync"; insp.src = ('https:' == document.location.protocol ? 'https' : 'http') + '://cdn.inspectlet.com/inspectlet.js?wid=814107329&r=' + Math.floor(new Date().getTime()/3600000); var x = document.getElementsByTagName('script')[0]; x.parentNode.insertBefore(insp, x); };
    setTimeout(ldinsp, 0);
    })();
    </script>
    <!-- End Inspectlet Asynchronous Code -->









  </body>
</html>
