<!doctype html>
<!--
  Minimal Mistakes Jekyll Theme 4.19.0 by Michael Rose
  Copyright 2013-2019 Michael Rose - mademistakes.com | @mmistakes
  Free for personal and commercial use under the MIT license
  https://github.com/mmistakes/minimal-mistakes/blob/master/LICENSE
-->
<html lang="en" class="no-js">
  <head>
    <meta charset="utf-8">

<!-- begin _includes/seo.html --><title>Multiple testing problem in Bioinformatics - Renesh Bedre</title>
<meta name="description" content="                               ">


  <meta name="author" content="Renesh Bedre">


<meta property="og:type" content="article">
<meta property="og:locale" content="en_US">
<meta property="og:site_name" content="Renesh Bedre">
<meta property="og:title" content="Multiple testing problem in Bioinformatics">
<meta property="og:url" content="https://reneshbedre.github.io//blog/mtest.html">


  <meta property="og:description" content="                               ">





  <meta name="twitter:site" content="@renbedre">
  <meta name="twitter:title" content="Multiple testing problem in Bioinformatics">
  <meta name="twitter:description" content="                               ">
  <meta name="twitter:url" content="https://reneshbedre.github.io//blog/mtest.html">

  
    <meta name="twitter:card" content="summary">
    
  

  



  <meta property="article:published_time" content="2018-12-16T10:19:20-06:00">





  

  


<link rel="canonical" href="https://reneshbedre.github.io//blog/mtest.html">




<script type="application/ld+json">
  {
    "@context": "https://schema.org",
    
      "@type": "Person",
      "name": "Renesh Bedre",
      "url": "https://reneshbedre.github.io//"
    
  }
</script>






<!-- end _includes/seo.html -->


<link href="/feed.xml" type="application/atom+xml" rel="alternate" title="Renesh Bedre Feed">

<!-- https://t.co/dKP3o1e -->
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<script>
  document.documentElement.className = document.documentElement.className.replace(/\bno-js\b/g, '') + ' js ';
</script>

<!-- For all browsers -->
<link rel="stylesheet" href="/assets/css/main.css">

<!--[if IE]>
  <style>
    /* old IE unsupported flexbox fixes */
    .greedy-nav .site-title {
      padding-right: 3em;
    }
    .greedy-nav button {
      position: absolute;
      top: 0;
      right: 0;
      height: 100%;
    }
  </style>
<![endif]-->



    <!-- start custom head snippets -->

<!-- insert favicons. use https://realfavicongenerator.net/ -->

<!-- end custom head snippets -->

  </head>

  <body class="layout--single wide">
    <nav class="skip-links">
  <h2 class="screen-reader-text">Skip links</h2>
  <ul>
    <li><a href="#site-nav" class="screen-reader-shortcut">Skip to primary navigation</a></li>
    <li><a href="#main" class="screen-reader-shortcut">Skip to content</a></li>
    <li><a href="#footer" class="screen-reader-shortcut">Skip to footer</a></li>
  </ul>
</nav>

    <!--[if lt IE 9]>
<div class="notice--danger align-center" style="margin: 0;">You are using an <strong>outdated</strong> browser. Please <a href="https://browsehappy.com/">upgrade your browser</a> to improve your experience.</div>
<![endif]-->

    

<div class="masthead">
  <div class="masthead__inner-wrap">
    <div class="masthead__menu">
      <nav id="site-nav" class="greedy-nav">
        
        <a class="site-title" href="/">
          Renesh Bedre
          
        </a>
        <ul class="visible-links"><li class="masthead__menu-item">
              <a href="/about/">About</a>
            </li><li class="masthead__menu-item">
              <a href="/blog/">Blog</a>
            </li><li class="masthead__menu-item">
              <a href="/pub/">Publications</a>
            </li><li class="masthead__menu-item">
              <a href="/news/">News</a>
            </li><li class="masthead__menu-item">
              <a href="/res/">Resources</a>
            </li></ul>
        
        <button class="greedy-nav__toggle hidden" type="button">
          <span class="visually-hidden">Toggle menu</span>
          <div class="navicon"></div>
        </button>
        <ul class="hidden-links hidden"></ul>
      </nav>
    </div>
  </div>
</div>


    <div class="initial-content">
      



<div id="main" role="main">
  
  <div class="sidebar sticky">
  


<div itemscope itemtype="https://schema.org/Person">

  
    <div class="author__avatar">
      
        <img src="/assets/images/me.png" alt="Renesh Bedre" itemprop="image">
      
    </div>
  

  <div class="author__content">
    
      <h3 class="author__name" itemprop="name">Renesh Bedre</h3>
    
    
      <div class="author__bio" itemprop="description">
        <p>Bioinformatics Scientist</p>

      </div>
    
  </div>

  <div class="author__urls-wrapper">
    <button class="btn btn--inverse">Follow</button>
    <ul class="author__urls social-icons">
      
        <li itemprop="homeLocation" itemscope itemtype="https://schema.org/Place">
          <i class="fas fa-fw fa-map-marker-alt" aria-hidden="true"></i> <span itemprop="name">Texas A&M University</span>
        </li>
      

      
        
          
        
          
            <li><a href="https://reneshbedre.github.io/" rel="nofollow noopener noreferrer"><i class="fas fa-fw fa-link" aria-hidden="true"></i> Website</a></li>
          
        
          
            <li><a href="https://twitter.com/renbedre" rel="nofollow noopener noreferrer"><i class="fab fa-fw fa-twitter-square" aria-hidden="true"></i> Twitter</a></li>
          
        
          
            <li><a href="https://www.linkedin.com/in/reneshbedre/" rel="nofollow noopener noreferrer"><i class="fab fa-linkedin-in" aria-hidden="true"></i> LinkedIn</a></li>
          
        
          
        
          
            <li><a href="https://github.com/reneshbedre" rel="nofollow noopener noreferrer"><i class="fab fa-fw fa-github" aria-hidden="true"></i> GitHub</a></li>
          
        
          
        
          
            <li><a href="assets/scipapers/Bedre_CV.pdf" rel="nofollow noopener noreferrer"><i class="ai ai-cv-square" aria-hidden="true"></i> CV</a></li>
          
        
      

      

      
        <li>
          <a href="mailto:reneshbe@gmail.com">
            <meta itemprop="email" content="reneshbe@gmail.com" />
            <i class="fas fa-fw fa-envelope-square" aria-hidden="true"></i> Email
          </a>
        </li>
      

      

      

      

      

      

      

      

      

      

      

      

      

      

      

      

      

      

      

      

      

      

      

      <!--
  <li>
    <a href="http://link-to-whatever-social-network.com/user/" itemprop="sameAs" rel="nofollow noopener noreferrer">
      <i class="fas fa-fw" aria-hidden="true"></i> Custom Social Profile Link
    </a>
  </li>
-->
    </ul>
  </div>
</div>

  
  </div>



  <article class="page" itemscope itemtype="https://schema.org/CreativeWork">
    <meta itemprop="headline" content="Multiple testing problem in Bioinformatics">
    <meta itemprop="description" content="    ">
    <meta itemprop="datePublished" content="2018-12-16T10:19:20-06:00">
    

    <div class="page__inner-wrap">
      
        <header>
          <h1 id="page-title" class="page__title" itemprop="headline">Multiple testing problem in Bioinformatics
</h1>

          <!-- change by me to add date and read time together -->
          <p class="page__meta">
          
           <!-- <p class="page__meta"><i class="far fa-clock" aria-hidden="true"></i> 




  5 minute read

</p>-->

            
              <i class="fa fa-fw fa-calendar" aria-hidden="true"></i> <time datetime="2018-12-16T10:19:20-06:00">December 16, 2018 </time>&emsp;
          
          &nbsp;<i class="far fa-clock" aria-hidden="true"></i>&nbsp;




  5 minute read


          

          </p>
        </header>
      

      <section class="page__content" itemprop="text">
        
        <script type="text/javascript" async="" src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js?config=TeX-MML-AM_CHTML"></script>

<p>
<link rel="stylesheet" href="https://maxcdn.bootstrapcdn.com/font-awesome/4.5.0/css/font-awesome.min.css" />

<style>
    .share-box a {
  display: inline-block;
  -webkit-box-shadow: 0 0 1px #777;
  box-shadow: 0 0 1px #777;
  padding: 5px 12px;
  margin-right: 5px;
  margin-bottom: 5px;
  text-decoration: none; }
  .share-box a:hover {
    text-decoration: none;
    -webkit-transition: background-color 200ms linear;
    -ms-transition: background-color 200ms linear;
    transition: background-color 200ms linear; }

.f {
  color: #3b5998; }
  .f:hover {
    color: #fff;
    background-color: #3b5998; }

.t {
  color: #4099FF; }
  .t:hover {
    color: #fff;
    background-color: #4099FF; }

.g {
  color: #d34836; }
  .g:hover {
    color: #fff;
    background-color: #d34836; }

.r {
  color: #ff5700; }
  .r:hover {
    color: #fff;
    background-color: #ff5700; }

.l {
  color: #0077b5; }
  .l:hover {
    color: #fff;
    background-color: #0077b5; }

.e {
  color: #444444; }
  .e:hover {
    color: #fff;
    background-color: #444444; }
</style>

<div class="share-box">
<!--<h5>Share this:</h5>-->
<a class="f" href="https://www.facebook.com/sharer/sharer.php?u=https://reneshbedre.github.io//blog/mtest.html" onclick="window.open(this.href, 'mywin',
'left=20,top=20,width=500,height=500,toolbar=1,resizable=0'); return false;"><i class="fab fa-facebook-f"></i></a>

<a class="t" href="https://twitter.com/intent/tweet?text=&amp;url=https://reneshbedre.github.io//blog/mtest.html" onclick="window.open(this.href, 'mywin',
'left=20,top=20,width=500,height=500,toolbar=1,resizable=0'); return false;"><i class="fab fa-twitter"></i></a>

  <a class="g" href="https://plus.google.com/share?url=https://reneshbedre.github.io//blog/mtest.html" onclick="window.open(this.href, 'mywin',
'left=20,top=20,width=500,height=500,toolbar=1,resizable=0'); return false;"><i class="fab fa-google-plus-square"></i></a>


<a class="r" href="http://www.reddit.com/submit?url=https://reneshbedre.github.io//blog/mtest.html" onclick="window.open(this.href, 'mywin',
'left=20,top=20,width=900,height=500,toolbar=1,resizable=0'); return false;"><i class="fab fa-reddit"></i></a>

<a class="l" href="https://www.linkedin.com/shareArticle?mini=true&amp;url=https://reneshbedre.github.io//blog/mtest.html" onclick="window.open(this.href, 'mywin',
'left=20,top=20,width=500,height=500,toolbar=1,resizable=0'); return false;"><i class="fab fa-linkedin-in"></i></a>

<a class="e" href="mailto:?subject=&amp;body=Check out this site https://reneshbedre.github.io//blog/mtest.html"><i class="fas fa-envelope-square"></i></a>

  <!--
    <a href="https://twitter.com/renbedre?ref_src=twsrc%5Etfw" class="twitter-follow-button" data-show-count="false">Follow @renbedre</a><script async src="https://platform.twitter.com/widgets.js" charset="utf-8"></script>
-->

</div>
</p>

<p><strong><span style="color:#33a8ff">What is multiple testing problem?</span></strong></p>
<ul>
  <li>In transcriptomics experiments, we analyze thousands of genes to identify significant differentially 
expressed genes (DEGs) between two groups (eg. normal vs. treated) which could alter biological mechanisms of a species in 
response to particular treatment</li>
  <li>In order to identify significant DEGs in terms of fold changes between two groups, it is necessary to apply multiple 
hypothesis testing (analysis of hypothesis across thousands of genes simultaneously). Each individual hypothesis 
testing for a single gene involves whether mean of two groups are different or not.</li>
  <li>In multiple hypothesis testing two kinds of errors that need to consider:<br />
<b>Type I error (False positive or chance of making mistake)</b>: Null hypothesis is rejected when it is fail to reject 
(Note: Type I error is also referred as P-value) <br />
<b>Type II error (False negative)</b>: Null hypothesis is fail to reject when it is reject</li>
  <li>For example, if we want to analyze 10000 genes to identify genes with significant fold changes between two gropus, 
the probability of getting significant results just due to by chance is 500 (5*10000/100) at a significance level of 
0.05 (5%) i.e 500 genes identified as false positive. The number of getting false positives will be more as we test large number of genes (generally in between 20K to 40K in human and plant 
species). Interpreting false positive results are very harmful as it will lead to wrong conclusions and therefore it is
essential to control false positive discovery.</li>
</ul>

<p><strong><span style="color:#33a8ff">How to minimize false discoveries?</span></strong></p>
<ul>
  <li>To control false discoveries from multiple hypothesis testing, it is imperative adjust significance level (α) to 
reduce the probability of getting Type I error (family-wise error rate or FWER)</li>
  <li>
    <p><b>Family-wise error rate (FWER)</b> is defined as a probability of getting at least one significant result just by chance.</p>

    <p>For example, If we perform 50 multiple tests at α=0.05, the probability of getting at least one significant 
result by chance is ~ 92% <br /></p>
    <p>
\(P(\text{at least one significant}) = 1−P(\text{no significant)}\) <br />
  \(\hspace{114 pt}                  = 1-(1-0.05)^{50} \sim 0.92\)
                        
</p>
    <p>In transcriptomics experiments, we analyze thousands of genes and therefore, probability of getting false positive 
results will be much higher.</p>
  </li>
  <li>FWER can be controlled by using <b>Bonferroni correction</b> where significance level (α) is adjusted to (α/n).          <br />
For example, If we perform 50 multiple tests at α=0.05, the probability of getting at least one significant 
result by chance is ~ 4% <br /> <br />
    <p>
The new adjusted &alpha; = 0.05/50 = 0.001 <br /> <br />
\(P(\text{at least one significant}) = 1−P(\text{no significant)}\) <br />
  \(\hspace{114 pt}                  = 1-(1-0.001)^{50} \sim 0.04\)
                        
</p>

    <p>Caution: Bonferroni correction is a highly conservative method. It can decrease the rate of false positives, but also 
extremely increases false negatives.</p>
  </li>
  <li>Another approach to control the false discoveries from multiple hypothesis testing is to control <b>false discovery rate 
 (FDR)</b>. FDR is defined as the proportion of false positives among the significant results. <b>Benjamini and Hochberg 
 (1995)</b> technique can be used to control FDR at a specific significance level. Benjamini and Hochberg is more
 suitable and recommended method for correcting P-values than Bonferroni correction in gene expression analysis.</li>
</ul>

<p><strong><span style="color:#33a8ff">When not to use multiple hypothesis testing corrections?</span></strong></p>

<p>P-value corrections methods for multiple hypothesis testing can increase the number of false negatives depending on the
type of methodology used. If false negative observations are very important or expensive, you should avoid correcting 
P-values during multiple hypothesis testing.</p>

<p>Below graph represents the probability of getting at least one significant result by chance with different numbers
tests at α = 0.05 and α = 0.01,</p>

<p align="center">
<img src="/assets/posts/multest/sign.png" width="600" />
</p>

<p><strong><span style="color:#33a8ff">Case study</span></strong></p>
<div class="language-python highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="c1"># I am using Python3
# load required packages
# make sure you have installed required packages
</span><span class="kn">from</span> <span class="nn">scipy.stats</span> <span class="kn">import</span> <span class="n">norm</span>
<span class="kn">import</span> <span class="nn">numpy</span> <span class="k">as</span> <span class="n">np</span>
<span class="kn">from</span> <span class="nn">statsmodels.stats.multitest</span> <span class="kn">import</span> <span class="n">multipletests</span>

<span class="c1"># generate 500 random numbers (observations) with mean 0 and standard deviation 1 for 
# multiple hypothesis testing
</span><span class="n">rand_num</span> <span class="o">=</span> <span class="n">norm</span><span class="o">.</span><span class="n">rvs</span><span class="p">(</span><span class="n">loc</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span> <span class="n">scale</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span> <span class="n">size</span><span class="o">=</span><span class="mi">500</span><span class="p">)</span>
<span class="c1"># now, get P-values for these observations using one-sided test
# Alternate hypothesis= observation value is larger than 0
# All observations should fail to reject null hypothesis and any obtained significant 
# P-value is just by a chance. We should expect ~25 (5*(500/100) significant P-value 
# (at 0.05 significance level) by a chance
</span><span class="n">pval</span> <span class="o">=</span> <span class="mi">1</span><span class="o">-</span><span class="n">norm</span><span class="o">.</span><span class="n">cdf</span><span class="p">(</span><span class="n">rand_num</span><span class="p">)</span> <span class="c1"># right-sided test
# count P-values &lt; 0.05
# output: 25
</span><span class="k">print</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">pval</span><span class="p">[</span><span class="n">np</span><span class="o">.</span><span class="n">where</span><span class="p">(</span><span class="n">pval</span><span class="o">&lt;</span><span class="mf">0.05</span><span class="p">)]))</span>

<span class="c1"># Bonferroni correction
# Get Bonferroni corrected P-value, which is 0.0001
</span><span class="n">bf_p</span> <span class="o">=</span> <span class="mf">0.05</span><span class="o">/</span><span class="mi">500</span> 
<span class="c1"># count P-values &lt; bf_p
# output: 0
</span><span class="k">print</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">pval</span><span class="p">[</span><span class="n">np</span><span class="o">.</span><span class="n">where</span><span class="p">(</span><span class="n">pval</span><span class="o">&lt;</span><span class="n">bf_p</span><span class="p">)]))</span>

<span class="c1"># Benjamini and Hochberg FDR at alpha=0.05
</span><span class="n">y</span><span class="o">=</span><span class="n">multipletests</span><span class="p">(</span><span class="n">pvals</span><span class="o">=</span><span class="n">pval</span><span class="p">,</span> <span class="n">alpha</span><span class="o">=</span><span class="mf">0.05</span><span class="p">,</span> <span class="n">method</span><span class="o">=</span><span class="s">"fdr_bh"</span><span class="p">)</span>
<span class="c1"># output: 0
</span><span class="k">print</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">y</span><span class="p">[</span><span class="mi">1</span><span class="p">][</span><span class="n">np</span><span class="o">.</span><span class="n">where</span><span class="p">(</span><span class="n">y</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span><span class="o">&lt;</span><span class="mf">0.05</span><span class="p">)]))</span>  <span class="c1"># y[1] returns corrected P-vals (array)
</span>
</code></pre></div></div>

<p>If you repeat above code, you should get significant P-values close to 25 which is also referred as Type I error rate 
(false positives) and calculated as (25/500=0.05). If you notice, the Type I error rate is equal to P-value 
significance level (0.05). After Bonferroni and Benjamini/Hochberg correction, we have obtained 0 false positives 
(Type I error rate=0/500=0).
<br /><br />
Bonferroni correction to control FWER: False positives = 0 <br />
Benjamini and Hochberg to control FDR: False positives = 0</p>

<div class="language-python highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="c1"># Now, generate 500 random numbers (observations) with mean 3 and standard deviation 1 for 
# multiple hypothesis testing
</span><span class="n">rand_num</span> <span class="o">=</span> <span class="n">norm</span><span class="o">.</span><span class="n">rvs</span><span class="p">(</span><span class="n">loc</span><span class="o">=</span><span class="mi">3</span><span class="p">,</span> <span class="n">scale</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span> <span class="n">size</span><span class="o">=</span><span class="mi">500</span><span class="p">)</span>
<span class="c1"># now, get P-values for these observations using one-sided test
# Alternate hypothesis= observation value is larger than 0
# All observations should reject null hypothesis (as we are performing cdf with mean=0) 
# and any obtained non-significant P-value is just by a chance
</span><span class="n">pval</span> <span class="o">=</span> <span class="mi">1</span><span class="o">-</span><span class="n">norm</span><span class="o">.</span><span class="n">cdf</span><span class="p">(</span><span class="n">rand_num</span><span class="p">)</span> <span class="c1"># right-sided test
# count P-values &lt; 0.05
# output: 460
</span><span class="k">print</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">pval</span><span class="p">[</span><span class="n">np</span><span class="o">.</span><span class="n">where</span><span class="p">(</span><span class="n">pval</span><span class="o">&lt;</span><span class="mf">0.05</span><span class="p">)]))</span>

<span class="c1"># Bonferroni correction
# Get Bonferroni corrected P-value, which is 0.0001
</span><span class="n">bf_p</span> <span class="o">=</span> <span class="mf">0.05</span><span class="o">/</span><span class="mi">500</span> 
<span class="c1"># count P-values &lt; bf_p
# output: 126
</span><span class="k">print</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">pval</span><span class="p">[</span><span class="n">np</span><span class="o">.</span><span class="n">where</span><span class="p">(</span><span class="n">pval</span><span class="o">&lt;</span><span class="n">bf_p</span><span class="p">)]))</span>

<span class="c1"># Benjamini and Hochberg FDR at alpha=0.05
</span><span class="n">y</span><span class="o">=</span><span class="n">multipletests</span><span class="p">(</span><span class="n">pvals</span><span class="o">=</span><span class="n">pval</span><span class="p">,</span> <span class="n">alpha</span><span class="o">=</span><span class="mf">0.05</span><span class="p">,</span> <span class="n">method</span><span class="o">=</span><span class="s">"fdr_bh"</span><span class="p">)</span>
<span class="c1"># output: 450
</span><span class="k">print</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">y</span><span class="p">[</span><span class="mi">1</span><span class="p">][</span><span class="n">np</span><span class="o">.</span><span class="n">where</span><span class="p">(</span><span class="n">y</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span><span class="o">&lt;</span><span class="mf">0.05</span><span class="p">)]))</span>  <span class="c1"># y[1] returns corrected P-vals (array)
</span></code></pre></div></div>

<p>If you repeat above code, you should get significant P-values close to 460. The Type II error rate (false negative) is
40/500 = 0.08. After Bonferroni correction, the rate of false negatives is tremendously increased (374/500=0.74), 
whereas with Benjamini/Hochberg correction, the rate of false negatives is 50/500 = 0.1.
<br /><br />
Bonferroni correction to control FWER: False negatives = 374 <br />
Benjamini and Hochberg to control FDR: False negatives = 50</p>

<p><strong><span style="color:#33a8ff">Reference</span></strong></p>
<ul>
  <li>Bonferroni CE. Il calcolo delle assicurazioni su gruppi di teste. Studi in onore del professore salvatore ortu 
carboni. 1935:13-60.</li>
  <li>Benjamini Y, Hochberg Y. Controlling the false discovery rate: a practical and powerful approach to multiple testing. 
Journal of the royal statistical society. Series B (Methodological). 1995 Jan 1:289-300.</li>
</ul>

<p><strong><span style="color:#33a8ff">How to cite?</span></strong></p>

<p>Bedre, R. Bioinformatics data analysis and visualization toolkit. GitHub repository, <a href="https://github.com/reneshbedre/bioinfokit">https://github.com/reneshbedre/bioinfokit</a></p>

<p><span style="color:#9e9696">If you have any questions, comments or recommendations, please email me at 
<b>reneshbe@gmail.com</b></span></p>

<p><span style="color:#9e9696"><i> Last updated: February 1, 2020</i> </span></p>

<p>
<link rel="stylesheet" href="https://maxcdn.bootstrapcdn.com/font-awesome/4.5.0/css/font-awesome.min.css" />

<style>
    .share-box a {
  display: inline-block;
  -webkit-box-shadow: 0 0 1px #777;
  box-shadow: 0 0 1px #777;
  padding: 5px 12px;
  margin-right: 5px;
  margin-bottom: 5px;
  text-decoration: none; }
  .share-box a:hover {
    text-decoration: none;
    -webkit-transition: background-color 200ms linear;
    -ms-transition: background-color 200ms linear;
    transition: background-color 200ms linear; }

.f {
  color: #3b5998; }
  .f:hover {
    color: #fff;
    background-color: #3b5998; }

.t {
  color: #4099FF; }
  .t:hover {
    color: #fff;
    background-color: #4099FF; }

.g {
  color: #d34836; }
  .g:hover {
    color: #fff;
    background-color: #d34836; }

.r {
  color: #ff5700; }
  .r:hover {
    color: #fff;
    background-color: #ff5700; }

.l {
  color: #0077b5; }
  .l:hover {
    color: #fff;
    background-color: #0077b5; }

.e {
  color: #444444; }
  .e:hover {
    color: #fff;
    background-color: #444444; }
</style>

<div class="share-box">
<!--<h5>Share this:</h5>-->
<a class="f" href="https://www.facebook.com/sharer/sharer.php?u=https://reneshbedre.github.io//blog/mtest.html" onclick="window.open(this.href, 'mywin',
'left=20,top=20,width=500,height=500,toolbar=1,resizable=0'); return false;"><i class="fab fa-facebook-f"></i></a>

<a class="t" href="https://twitter.com/intent/tweet?text=&amp;url=https://reneshbedre.github.io//blog/mtest.html" onclick="window.open(this.href, 'mywin',
'left=20,top=20,width=500,height=500,toolbar=1,resizable=0'); return false;"><i class="fab fa-twitter"></i></a>

  <a class="g" href="https://plus.google.com/share?url=https://reneshbedre.github.io//blog/mtest.html" onclick="window.open(this.href, 'mywin',
'left=20,top=20,width=500,height=500,toolbar=1,resizable=0'); return false;"><i class="fab fa-google-plus-square"></i></a>


<a class="r" href="http://www.reddit.com/submit?url=https://reneshbedre.github.io//blog/mtest.html" onclick="window.open(this.href, 'mywin',
'left=20,top=20,width=900,height=500,toolbar=1,resizable=0'); return false;"><i class="fab fa-reddit"></i></a>

<a class="l" href="https://www.linkedin.com/shareArticle?mini=true&amp;url=https://reneshbedre.github.io//blog/mtest.html" onclick="window.open(this.href, 'mywin',
'left=20,top=20,width=500,height=500,toolbar=1,resizable=0'); return false;"><i class="fab fa-linkedin-in"></i></a>

<a class="e" href="mailto:?subject=&amp;body=Check out this site https://reneshbedre.github.io//blog/mtest.html"><i class="fas fa-envelope-square"></i></a>

  <!--
    <a href="https://twitter.com/renbedre?ref_src=twsrc%5Etfw" class="twitter-follow-button" data-show-count="false">Follow @renbedre</a><script async src="https://platform.twitter.com/widgets.js" charset="utf-8"></script>
-->

</div>
</p>

        
      </section>




      <footer class="page__meta">
        
        


        <!-- commented to remove updated in footer section
        
          <p class="page__date"><strong><i class="fas fa-fw fa-calendar-alt" aria-hidden="true"></i> Updated:</strong> <time datetime="2018-12-16T10:19:20-06:00">December 16, 2018</time></p>
        
        -->
      </footer>

      <section class="page__share">
  
    <h4 class="page__share-title">Share on</h4>
  

  <a href="https://twitter.com/intent/tweet?via=renbedre&text=Multiple+testing+problem+in+Bioinformatics%20https%3A%2F%2Freneshbedre.github.io%2F%2Fblog%2Fmtest.html" class="btn btn--twitter" onclick="window.open(this.href, 'window', 'left=20,top=20,width=500,height=500,toolbar=1,resizable=0'); return false;" title="Share on Twitter"><i class="fab fa-fw fa-twitter" aria-hidden="true"></i><span> Twitter</span></a>

  <a href="https://www.facebook.com/sharer/sharer.php?u=https%3A%2F%2Freneshbedre.github.io%2F%2Fblog%2Fmtest.html" class="btn btn--facebook" onclick="window.open(this.href, 'window', 'left=20,top=20,width=500,height=500,toolbar=1,resizable=0'); return false;" title="Share on Facebook"><i class="fab fa-fw fa-facebook" aria-hidden="true"></i><span> Facebook</span></a>

  <a href="https://www.linkedin.com/shareArticle?mini=true&url=https%3A%2F%2Freneshbedre.github.io%2F%2Fblog%2Fmtest.html" class="btn btn--linkedin" onclick="window.open(this.href, 'window', 'left=20,top=20,width=500,height=500,toolbar=1,resizable=0'); return false;" title="Share on LinkedIn"><i class="fab fa-fw fa-linkedin" aria-hidden="true"></i><span> LinkedIn</span></a>
</section>


      
  <nav class="pagination">
    
      <a href="/blog/fqutil.html" class="pagination--pager" title="How to use NCBI SRA Toolkit effectively?
">Previous</a>
    
    
      <a href="/blog/corr.html" class="pagination--pager" title="Correlation analysis in Python
">Next</a>
    
  </nav>

    </div>

    
  </article>

  
  
    <div class="page__related">
      <h4 class="page__related-title">You may also enjoy</h4>
      <div class="grid__wrapper">
        
          



<div class="grid__item">
  <article class="archive__item" itemscope itemtype="https://schema.org/CreativeWork">
    
    <h2 class="archive__item-title" itemprop="headline">
      
        <a href="/blog/howtoinstall.html" rel="permalink">How to install bioinfokit
</a>
      
    </h2>
    
      <p class="page__meta"><i class="far fa-clock" aria-hidden="true"></i> 




  less than 1 minute read

</p>
    
    <p class="archive__item-excerpt" itemprop="description">










  








  




</p>
  </article>
</div>

        
          



<div class="grid__item">
  <article class="archive__item" itemscope itemtype="https://schema.org/CreativeWork">
    
    <h2 class="archive__item-title" itemprop="headline">
      
        <a href="/blog/fqqualfmt.html" rel="permalink">FASTQ Sequence Quality Format
</a>
      
    </h2>
    
      <p class="page__meta"><i class="far fa-clock" aria-hidden="true"></i> 




  3 minute read

</p>
    
    <p class="archive__item-excerpt" itemprop="description">

</p>
  </article>
</div>

        
          



<div class="grid__item">
  <article class="archive__item" itemscope itemtype="https://schema.org/CreativeWork">
    
    <h2 class="archive__item-title" itemprop="headline">
      
        <a href="/blog/bardot.html" rel="permalink">Bar graph with overlapping dots (Bar-dot) plot
</a>
      
    </h2>
    
      <p class="page__meta"><i class="far fa-clock" aria-hidden="true"></i> 




  1 minute read

</p>
    
    <p class="archive__item-excerpt" itemprop="description">










  








  




</p>
  </article>
</div>

        
          



<div class="grid__item">
  <article class="archive__item" itemscope itemtype="https://schema.org/CreativeWork">
    
    <h2 class="archive__item-title" itemprop="headline">
      
        <a href="/blog/manhat.html" rel="permalink">Manhattan plot
</a>
      
    </h2>
    
      <p class="page__meta"><i class="far fa-clock" aria-hidden="true"></i> 




  2 minute read

</p>
    
    <p class="archive__item-excerpt" itemprop="description">










  








  




</p>
  </article>
</div>

        
      </div>
    </div>
  
  
</div>

    </div>

    

    <div id="footer" class="page__footer">
      <footer>
        <!-- start custom footer snippets -->

<!-- end custom footer snippets -->
        <link rel="stylesheet" href="https://cdn.rawgit.com/jpswalsh/academicons/master/css/academicons.min.css">

<div class="page__footer-follow">
  <ul class="social-icons">
    
      <li><strong>Follow:</strong></li>
    

    
      
        
          <li><a href="https://twitter.com/renbedre" rel="nofollow noopener noreferrer"><i class="fab fa-fw fa-twitter-square" aria-hidden="true"></i> Twitter</a></li>
        
      
        
      
        
          <li><a href="https://github.com/reneshbedre" rel="nofollow noopener noreferrer"><i class="fab fa-fw fa-github" aria-hidden="true"></i> GitHub</a></li>
        
      
        
      
        
      
        
      
        
          <li><a href="https://scholar.google.com/citations?user=6pv2AiMAAAAJ&hl=en" rel="nofollow noopener noreferrer"><i class="ai ai-google-scholar-square" aria-hidden="true"></i> Google Scholar</a></li>
        
      
        
          <li><a href="https://scholar.google.com/citations?user=6pv2AiMAAAAJ&hl=en" rel="nofollow noopener noreferrer"><i class="ai ai-researchgate-square" aria-hidden="true"></i> ResearchGate</a></li>
        
      
    

    <li><a href="/feed.xml"><i class="fas fa-fw fa-rss-square" aria-hidden="true"></i> Feed</a></li>
  </ul>
</div>

<div class="page__footer-copyright">&copy; 2020 Renesh Bedre. Powered by <a href="https://jekyllrb.com" rel="nofollow">Jekyll</a> &amp; <a href="https://mademistakes.com/work/minimal-mistakes-jekyll-theme/" rel="nofollow">Minimal Mistakes</a>.</div>

      </footer>
    </div>

    
  <script src="/assets/js/main.min.js"></script>
  <script src="https://kit.fontawesome.com/4eee35f757.js"></script>







  <!-- Global site tag (gtag.js) - Google Analytics -->
<script async src="https://www.googletagmanager.com/gtag/js?id=UA-115112853-1"></script>
<script>
  window.dataLayer = window.dataLayer || [];
  function gtag(){dataLayer.push(arguments);}
  gtag('js', new Date());

  gtag('config', 'UA-115112853-1', { 'anonymize_ip': false});
</script>


<!-- Begin Inspectlet Asynchronous Code -->
    <script type="text/javascript">
    (function() {
    window.__insp = window.__insp || [];
    __insp.push(['wid', 814107329]);
    var ldinsp = function(){
    if(typeof window.__inspld != "undefined") return; window.__inspld = 1; var insp = document.createElement('script'); insp.type = 'text/javascript'; insp.async = true; insp.id = "inspsync"; insp.src = ('https:' == document.location.protocol ? 'https' : 'http') + '://cdn.inspectlet.com/inspectlet.js?wid=814107329&r=' + Math.floor(new Date().getTime()/3600000); var x = document.getElementsByTagName('script')[0]; x.parentNode.insertBefore(insp, x); };
    setTimeout(ldinsp, 0);
    })();
    </script>
    <!-- End Inspectlet Asynchronous Code -->









  </body>
</html>
