<!doctype html>
<!--
  Minimal Mistakes Jekyll Theme 4.19.0 by Michael Rose
  Copyright 2013-2019 Michael Rose - mademistakes.com | @mmistakes
  Free for personal and commercial use under the MIT license
  https://github.com/mmistakes/minimal-mistakes/blob/master/LICENSE
-->
<html lang="en" class="no-js">
  <head>
    <meta charset="utf-8">

<!-- begin _includes/seo.html --><title>Principal component analysis (PCA) analysis and visualization using Python - Renesh Bedre</title>
<meta name="description" content="                             ">


  <meta name="author" content="Renesh Bedre">


<meta property="og:type" content="article">
<meta property="og:locale" content="en_US">
<meta property="og:site_name" content="Renesh Bedre">
<meta property="og:title" content="Principal component analysis (PCA) analysis and visualization using Python">
<meta property="og:url" content="https://reneshbedre.github.io//blog/pca_3d.html">


  <meta property="og:description" content="                             ">





  <meta name="twitter:site" content="@renbedre">
  <meta name="twitter:title" content="Principal component analysis (PCA) analysis and visualization using Python">
  <meta name="twitter:description" content="                             ">
  <meta name="twitter:url" content="https://reneshbedre.github.io//blog/pca_3d.html">

  
    <meta name="twitter:card" content="summary">
    
  

  



  <meta property="article:published_time" content="2018-03-30T01:18:08-05:00">





  

  


<link rel="canonical" href="https://reneshbedre.github.io//blog/pca_3d.html">




<script type="application/ld+json">
  {
    "@context": "https://schema.org",
    
      "@type": "Person",
      "name": "Renesh Bedre",
      "url": "https://reneshbedre.github.io//"
    
  }
</script>






<!-- end _includes/seo.html -->


<link href="/feed.xml" type="application/atom+xml" rel="alternate" title="Renesh Bedre Feed">

<!-- https://t.co/dKP3o1e -->
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<script>
  document.documentElement.className = document.documentElement.className.replace(/\bno-js\b/g, '') + ' js ';
</script>

<!-- For all browsers -->
<link rel="stylesheet" href="/assets/css/main.css">

<!--[if IE]>
  <style>
    /* old IE unsupported flexbox fixes */
    .greedy-nav .site-title {
      padding-right: 3em;
    }
    .greedy-nav button {
      position: absolute;
      top: 0;
      right: 0;
      height: 100%;
    }
  </style>
<![endif]-->



    <!-- start custom head snippets -->

<!-- insert favicons. use https://realfavicongenerator.net/ -->

<!-- end custom head snippets -->

  </head>

  <body class="layout--single wide">
    <nav class="skip-links">
  <h2 class="screen-reader-text">Skip links</h2>
  <ul>
    <li><a href="#site-nav" class="screen-reader-shortcut">Skip to primary navigation</a></li>
    <li><a href="#main" class="screen-reader-shortcut">Skip to content</a></li>
    <li><a href="#footer" class="screen-reader-shortcut">Skip to footer</a></li>
  </ul>
</nav>

    <!--[if lt IE 9]>
<div class="notice--danger align-center" style="margin: 0;">You are using an <strong>outdated</strong> browser. Please <a href="https://browsehappy.com/">upgrade your browser</a> to improve your experience.</div>
<![endif]-->

    

<div class="masthead">
  <div class="masthead__inner-wrap">
    <div class="masthead__menu">
      <nav id="site-nav" class="greedy-nav">
        
        <a class="site-title" href="/">
          Renesh Bedre
          
        </a>
        <ul class="visible-links"><li class="masthead__menu-item">
              <a href="/about/">About</a>
            </li><li class="masthead__menu-item">
              <a href="/blog/">Blog</a>
            </li><li class="masthead__menu-item">
              <a href="/pub/">Publications</a>
            </li><li class="masthead__menu-item">
              <a href="/news/">News</a>
            </li><li class="masthead__menu-item">
              <a href="/res/">Resources</a>
            </li></ul>
        
        <button class="greedy-nav__toggle hidden" type="button">
          <span class="visually-hidden">Toggle menu</span>
          <div class="navicon"></div>
        </button>
        <ul class="hidden-links hidden"></ul>
      </nav>
    </div>
  </div>
</div>


    <div class="initial-content">
      



<div id="main" role="main">
  
  <div class="sidebar sticky">
  


<div itemscope itemtype="https://schema.org/Person">

  
    <div class="author__avatar">
      
        <img src="/assets/images/me.png" alt="Renesh Bedre" itemprop="image">
      
    </div>
  

  <div class="author__content">
    
      <h3 class="author__name" itemprop="name">Renesh Bedre</h3>
    
    
      <div class="author__bio" itemprop="description">
        <p>Bioinformatics Scientist</p>

      </div>
    
  </div>

  <div class="author__urls-wrapper">
    <button class="btn btn--inverse">Follow</button>
    <ul class="author__urls social-icons">
      
        <li itemprop="homeLocation" itemscope itemtype="https://schema.org/Place">
          <i class="fas fa-fw fa-map-marker-alt" aria-hidden="true"></i> <span itemprop="name">Texas A&M University</span>
        </li>
      

      
        
          
        
          
            <li><a href="https://reneshbedre.github.io/" rel="nofollow noopener noreferrer"><i class="fas fa-fw fa-link" aria-hidden="true"></i> Website</a></li>
          
        
          
            <li><a href="https://twitter.com/renbedre" rel="nofollow noopener noreferrer"><i class="fab fa-fw fa-twitter-square" aria-hidden="true"></i> Twitter</a></li>
          
        
          
            <li><a href="https://www.linkedin.com/in/reneshbedre/" rel="nofollow noopener noreferrer"><i class="fab fa-linkedin-in" aria-hidden="true"></i> LinkedIn</a></li>
          
        
          
        
          
            <li><a href="https://github.com/reneshbedre" rel="nofollow noopener noreferrer"><i class="fab fa-fw fa-github" aria-hidden="true"></i> GitHub</a></li>
          
        
          
        
          
            <li><a href="assets/scipapers/Bedre_CV.pdf" rel="nofollow noopener noreferrer"><i class="ai ai-cv-square" aria-hidden="true"></i> CV</a></li>
          
        
      

      

      
        <li>
          <a href="mailto:reneshbe@gmail.com">
            <meta itemprop="email" content="reneshbe@gmail.com" />
            <i class="fas fa-fw fa-envelope-square" aria-hidden="true"></i> Email
          </a>
        </li>
      

      

      

      

      

      

      

      

      

      

      

      

      

      

      

      

      

      

      

      

      

      

      

      <!--
  <li>
    <a href="http://link-to-whatever-social-network.com/user/" itemprop="sameAs" rel="nofollow noopener noreferrer">
      <i class="fas fa-fw" aria-hidden="true"></i> Custom Social Profile Link
    </a>
  </li>
-->
    </ul>
  </div>
</div>

  
  </div>



  <article class="page" itemscope itemtype="https://schema.org/CreativeWork">
    <meta itemprop="headline" content="Principal component analysis (PCA) analysis and visualization using Python">
    <meta itemprop="description" content="    ">
    <meta itemprop="datePublished" content="2018-03-30T01:18:08-05:00">
    

    <div class="page__inner-wrap">
      
        <header>
          <h1 id="page-title" class="page__title" itemprop="headline">Principal component analysis (PCA) analysis and visualization using Python
</h1>

          <!-- change by me to add date and read time together -->
          <p class="page__meta">
          
           <!-- <p class="page__meta"><i class="far fa-clock" aria-hidden="true"></i> 




  2 minute read

</p>-->

            
              <i class="fa fa-fw fa-calendar" aria-hidden="true"></i> <time datetime="2018-03-30T01:18:08-05:00">March 30, 2018 </time>&emsp;
          
          &nbsp;<i class="far fa-clock" aria-hidden="true"></i>&nbsp;




  2 minute read


          

          </p>
        </header>
      

      <section class="page__content" itemprop="text">
        
        <p>
<link rel="stylesheet" href="https://maxcdn.bootstrapcdn.com/font-awesome/4.5.0/css/font-awesome.min.css" />

<style>
    .share-box a {
  display: inline-block;
  -webkit-box-shadow: 0 0 1px #777;
  box-shadow: 0 0 1px #777;
  padding: 5px 12px;
  margin-right: 5px;
  margin-bottom: 5px;
  text-decoration: none; }
  .share-box a:hover {
    text-decoration: none;
    -webkit-transition: background-color 200ms linear;
    -ms-transition: background-color 200ms linear;
    transition: background-color 200ms linear; }

.f {
  color: #3b5998; }
  .f:hover {
    color: #fff;
    background-color: #3b5998; }

.t {
  color: #4099FF; }
  .t:hover {
    color: #fff;
    background-color: #4099FF; }

.g {
  color: #d34836; }
  .g:hover {
    color: #fff;
    background-color: #d34836; }

.r {
  color: #ff5700; }
  .r:hover {
    color: #fff;
    background-color: #ff5700; }

.l {
  color: #0077b5; }
  .l:hover {
    color: #fff;
    background-color: #0077b5; }

.e {
  color: #444444; }
  .e:hover {
    color: #fff;
    background-color: #444444; }
</style>

<div class="share-box">
<!--<h5>Share this:</h5>-->
<a class="f" href="https://www.facebook.com/sharer/sharer.php?u=https://reneshbedre.github.io//blog/pca_3d.html" onclick="window.open(this.href, 'mywin',
'left=20,top=20,width=500,height=500,toolbar=1,resizable=0'); return false;"><i class="fab fa-facebook-f"></i></a>

<a class="t" href="https://twitter.com/intent/tweet?text=&amp;url=https://reneshbedre.github.io//blog/pca_3d.html" onclick="window.open(this.href, 'mywin',
'left=20,top=20,width=500,height=500,toolbar=1,resizable=0'); return false;"><i class="fab fa-twitter"></i></a>

  <a class="g" href="https://plus.google.com/share?url=https://reneshbedre.github.io//blog/pca_3d.html" onclick="window.open(this.href, 'mywin',
'left=20,top=20,width=500,height=500,toolbar=1,resizable=0'); return false;"><i class="fab fa-google-plus-square"></i></a>


<a class="r" href="http://www.reddit.com/submit?url=https://reneshbedre.github.io//blog/pca_3d.html" onclick="window.open(this.href, 'mywin',
'left=20,top=20,width=900,height=500,toolbar=1,resizable=0'); return false;"><i class="fab fa-reddit"></i></a>

<a class="l" href="https://www.linkedin.com/shareArticle?mini=true&amp;url=https://reneshbedre.github.io//blog/pca_3d.html" onclick="window.open(this.href, 'mywin',
'left=20,top=20,width=500,height=500,toolbar=1,resizable=0'); return false;"><i class="fab fa-linkedin-in"></i></a>

<a class="e" href="mailto:?subject=&amp;body=Check out this site https://reneshbedre.github.io//blog/pca_3d.html"><i class="fas fa-envelope-square"></i></a>

  <!--
    <a href="https://twitter.com/renbedre?ref_src=twsrc%5Etfw" class="twitter-follow-button" data-show-count="false">Follow @renbedre</a><script async src="https://platform.twitter.com/widgets.js" charset="utf-8"></script>
-->

</div>
</p>

<p>Transcriptomics  experiments such as RNA-seq allows researchers to study large numbers of genes across multiple treatment
conditions simultaneously. There have been various analytical methods has been used to deduce underlying biological
information from gene expression data such as heatmaps, clustering, PCA, Venn diagrams, gene networks and so on.</p>

<p>PCA is a classical multivariate (unconstrained ordination) statistical method that used to interpret the variation in dataset usually when the dataset
contains a large number of variables to study. PCA does this by reducing the dimensionality of the dataset by transforming old variable into a new
set of variables called principal component (PC), which are easy to visualize and summarise the features of datasets.
In gene expression datasets, variables can be treatment comparisons, gene expression counts across various conditions,
individual samples etc.</p>

<p>For example, when datasets contain 10 variables (10D), it is arduous to visualize them at the same time
(you may have to do 45 pairwise comparisons to interpret dataset effectively). PCA transforms them into a new set of
variables (PCs) with top PCs having highest variation. Most of time first 3 PCs contribute most of the variance. These
top 2 or 3 PCs can be plotted easily and summarize and/or clusters the features of all 10 variables.</p>

<p>For simplicity and understanding, here I am doing PCA analysis and visualization on small dataset (4 variables). But, 
once you understood, you can follow similar steps to run PCA analysis and visualization on large gene expression 
<a href="/assets/posts/pca/pca_example_dataset.csv">dataset</a>. (This dataset contains 13324 genes and 18 variables (A to R). 
These variables represent the log2 expression fold changes between different treatments)</p>

<div class="language-python highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="c1"># We will use `bioinfokit` for PCA analysis
# download and install bioinfokit (Tested on Linux, Mac, Windows) 
# Read documentation at https://github.com/reneshbedre/bioinfokit
</span><span class="n">git</span> <span class="n">clone</span> <span class="n">https</span><span class="p">:</span><span class="o">//</span><span class="n">github</span><span class="o">.</span><span class="n">com</span><span class="o">/</span><span class="n">reneshbedre</span><span class="o">/</span><span class="n">bioinfokit</span><span class="o">.</span><span class="n">git</span>
<span class="n">cd</span> <span class="n">bioinfokit</span>
<span class="n">python3</span> <span class="n">setup</span><span class="o">.</span><span class="n">py</span> <span class="n">install</span>
</code></pre></div></div>

<div class="language-python highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="c1"># you can use interactive python console, jupyter or python code
# I am using interactive python console (Python 3.6)
# Read documentation at https://github.com/reneshbedre/bioinfokit
</span><span class="o">&gt;&gt;&gt;</span><span class="kn">from</span> <span class="nn">bioinfokit</span> <span class="kn">import</span> <span class="n">analys</span>
<span class="o">&gt;&gt;&gt;</span><span class="kn">import</span> <span class="nn">pandas</span> <span class="k">as</span> <span class="n">pd</span>
<span class="c1"># load data file
</span><span class="o">&gt;&gt;&gt;</span><span class="n">d</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="s">"https://reneshbedre.github.io/myfiles/bioinfokit_data/pca_small_dataset.csv"</span><span class="p">)</span>
<span class="c1">#view first few rows
</span><span class="o">&gt;&gt;&gt;</span><span class="n">d</span><span class="o">.</span><span class="n">head</span><span class="p">()</span>
 	<span class="n">rowid</span> 	<span class="n">A</span> 	<span class="n">B</span> 	<span class="n">C</span> 	<span class="n">D</span>
<span class="mi">0</span> 	<span class="n">row1</span> 	<span class="mi">25</span> 	<span class="mi">45</span> 	<span class="mi">30</span> 	<span class="mi">54</span>
<span class="mi">1</span> 	<span class="n">row2</span> 	<span class="mi">30</span> 	<span class="mi">55</span> 	<span class="mi">29</span> 	<span class="mi">60</span>
<span class="mi">2</span> 	<span class="n">row3</span> 	<span class="mi">28</span> 	<span class="mi">29</span> 	<span class="mi">33</span> 	<span class="mi">51</span>
<span class="mi">3</span> 	<span class="n">row4</span> 	<span class="mi">36</span> 	<span class="mi">56</span> 	<span class="mi">37</span> 	<span class="mi">62</span>
<span class="mi">4</span> 	<span class="n">row5</span> 	<span class="mi">29</span> 	<span class="mi">40</span> 	<span class="mi">27</span> 	<span class="mi">73</span>
<span class="c1"># make sure the dataframe should not have first identifier columns. 
# if table has identifier columns, drop it using d=d.drop('column_name', 1) 
</span><span class="o">&gt;&gt;&gt;</span><span class="n">d</span> <span class="o">=</span> <span class="n">d</span><span class="o">.</span><span class="n">drop</span><span class="p">(</span><span class="s">"rowid"</span><span class="p">,</span> <span class="mi">1</span><span class="p">)</span>
<span class="c1"># now, perform pca analysis and visualization
</span><span class="o">&gt;&gt;&gt;</span><span class="n">analys</span><span class="o">.</span><span class="n">pca</span><span class="p">(</span><span class="n">table</span><span class="o">=</span><span class="n">d</span><span class="p">)</span>

<span class="n">Component</span> <span class="n">summary</span>

                             <span class="n">PC1</span>       <span class="n">PC2</span>       <span class="n">PC3</span>       <span class="n">PC4</span>
<span class="n">Proportion</span> <span class="n">of</span> <span class="n">Variance</span>  <span class="mf">0.619541</span>  <span class="mf">0.291498</span>  <span class="mf">0.085472</span>  <span class="mf">0.003489</span>
<span class="n">Cumulative</span> <span class="n">proportion</span>   <span class="mf">0.619541</span>  <span class="mf">0.911039</span>  <span class="mf">0.996511</span>  <span class="mf">1.000000</span>

<span class="n">Loadings</span>

      <span class="n">PC1</span>       <span class="n">PC2</span>       <span class="n">PC3</span>       <span class="n">PC4</span>
<span class="n">A</span> <span class="o">-</span><span class="mf">0.222185</span> <span class="o">-</span><span class="mf">0.020970</span>  <span class="mf">0.671764</span>  <span class="mf">0.706348</span>
<span class="n">B</span> <span class="o">-</span><span class="mf">0.905969</span> <span class="o">-</span><span class="mf">0.336097</span> <span class="o">-</span><span class="mf">0.251237</span> <span class="o">-</span><span class="mf">0.056019</span>
<span class="n">C</span> <span class="o">-</span><span class="mf">0.042114</span> <span class="o">-</span><span class="mf">0.285112</span>  <span class="mf">0.682941</span> <span class="o">-</span><span class="mf">0.671215</span>
<span class="n">D</span> <span class="o">-</span><span class="mf">0.357882</span>  <span class="mf">0.897391</span>  <span class="mf">0.138581</span> <span class="o">-</span><span class="mf">0.217728</span>
<span class="c1"># plot will be saved in same directory (screeplot.png, pcaplot_2d.png, and pcaplot_3d.png)
</span></code></pre></div></div>

<p>Generated Scree plot,</p>

<p><img src="https://reneshbedre.github.io//assets/posts/pca/screeplot.png" alt="screenshot" /></p>

<p>Generated PCA plot (2 PCs) plot,</p>

<p><img src="https://reneshbedre.github.io//assets/posts/pca/pcaplot_2d.png" alt="screenshot" /></p>

<p>Generated PCA plot (3 PCs) plot,</p>

<p><img src="https://reneshbedre.github.io//assets/posts/pca/pcaplot_3d.png" alt="screenshot" /></p>

<!--

In this tutorial, I will discuss in detail how to perform PCA analysis, visualization and interpretation using sample gene
expression data. For PCA analysis and visualization, I will use [prcomp](https://www.rdocumentation.org/packages/stats/versions/3.4.3/topics/prcomp)
and [scatter3d](https://www.rdocumentation.org/packages/car/versions/2.1-6/topics/scatter3d) R packages.

<p>Sample dataset used in this tutorial <a href="/myfiles/pca_example_dataset.csv">dataset</a>. This sample
gene expression dataset contains 13324 genes and 18 variables (A to R). These variables represent the log2 expression
fold changes between different treatments.</p>

First, let's load the R libraries needed for analysis. You can install using `install.packages("package")` or using
`bioclite()` [How to install R packages](https://www.r-bloggers.com/installing-r-packages/)

```r
# load libraries
>library(gplots)
>library(plot3D)
>library(factoextra)

# import data
>data_table = read.csv("pca_example_dataset.csv", header=TRUE, row.names=1)

# convert table to matrix
>data_table_mat = data.matrix(data_table)

# remove all rows where all columns have 0 values; this is optional
>data_table_mat = data_table_mat[!apply(data_table_mat, 1,
                  function(x) all(x == 0)), ]

# check the dimension of data; this is optional
>dim(data_table_mat)

# perform PCA analysis
# As the gene expression data represents log2 fold change for each variable,
# I did not use data scaling. But if your data is not uniform set scale. to TRUE
>pca = prcomp(data_table_mat, center = TRUE, scale. = FALSE)
>summary(pca)

Importance of components:
                          PC1    PC2    PC3    PC4    PC5    PC6     PC7     PC8     PC9    PC10    PC11    PC12    PC13    PC14    PC15    PC16    PC17    PC18
Standard deviation     1.9455 1.4881 1.0065 0.8670 0.8059 0.5280 0.44982 0.29553 0.20272 0.16985 0.16136 0.15739 0.15648 0.14507 0.14080 0.12359 0.10881 0.07024
Proportion of Variance 0.4114 0.2407 0.1101 0.0817 0.0706 0.0303 0.02199 0.00949 0.00447 0.00314 0.00283 0.00269 0.00266 0.00229 0.00215 0.00166 0.00129 0.00054
Cumulative Proportion  0.4114 0.6521 0.7622 0.8439 0.9145 0.9448 0.96679 0.97629 0.98075 0.98389 0.98672 0.98941 0.99207 0.99436 0.99652 0.99818 0.99946 1.00000

> print(pca)
Standard deviations:
 [1] 1.9454740 1.4880623 1.0064682 0.8669863 0.8059281 0.5279759 0.4498158 0.2955292 0.2027190 0.1698526 0.1613629 0.1573882 0.1564750 0.1450662 0.1407983 0.1235900
[17] 0.1088081 0.0702448

Rotation:
            PC1           PC2           PC3           PC4           PC5           PC6           PC7           PC8          PC9        PC10         PC11          PC12
A -1.343097e-02  6.323911e-02 -0.0691930665 -0.0251485010  2.648116e-01 -0.0068880765  0.7378477434 -2.797293e-02 -0.013313448  0.45133597 -0.047667148  3.798500e-01
B  9.944610e-04  2.052127e-03 -0.0010546396 -0.0014434767 -3.881107e-04  0.0119064932  0.0043613575  6.916949e-02  0.581932245  0.03253789 -0.028603243  4.767295e-02
C -8.140684e-02  1.399650e-01 -0.1785363909 -0.1751961080  9.017922e-01  0.0181988685 -0.1205413085 -6.525849e-05  0.003943627 -0.21588497  0.016013714 -1.738292e-01
D  1.077459e-03  5.397563e-03 -0.0017227414 -0.0039767523  1.267614e-03  0.0026945290  0.0033614634  1.299098e-01  0.698601169  0.02934107 -0.005190133  8.856802e-03
E -2.649007e-02  1.596509e-02 -0.0188695088 -0.0418816412  1.298501e-01  0.0055685428 -0.6620367699  3.339199e-03 -0.014262923  0.54107576 -0.051299478  4.548146e-01
F  1.037250e-05 -2.405279e-04 -0.0006820423 -0.0006352852 -2.638620e-04  0.0002915963 -0.0031961689  9.196993e-03  0.033868485 -0.00936920 -0.011872259 -3.157451e-02
G -2.386615e-01  6.197818e-01  0.1332794399  0.5103241553  1.868681e-03 -0.0003165033 -0.0283642481 -3.434216e-03  0.006738578  0.10670154 -0.005890525 -5.706278e-03
H  1.391363e-04  7.681416e-05  0.0037942991  0.0009830673 -1.355079e-05  0.0180411335  0.0011590960 -8.100548e-02 -0.244308428  0.11649427  0.371923927  1.430474e-01
I -3.025029e-01  6.429415e-01  0.1339030619 -0.4223749927 -1.832009e-01  0.0015723094  0.0152779069  2.410008e-03 -0.011718735 -0.11804666  0.003760751  1.711777e-04
J  2.560176e-04 -2.473587e-03  0.0009164680  0.0065768346  3.258796e-03 -0.0004740609  0.0007920751 -1.553285e-02 -0.240716220 -0.04818989  0.160745710 -1.576680e-02
K -4.744437e-02  2.908348e-03 -0.0041924433 -0.7225472267 -1.425278e-01  0.0057912403  0.0134547203 -4.342178e-03  0.004161067  0.14724539 -0.005147585  4.745356e-06
L  4.715754e-05  4.907378e-04  0.0003933394 -0.0005349674  1.659020e-03  0.0052901093 -0.0010713225  1.314869e-02  0.014867245 -0.06628550  0.085932013 -5.681064e-02
M -1.638196e-01  6.155082e-02 -0.7333155095  0.0443475349 -1.597227e-01 -0.0132622111 -0.0110459916 -6.070002e-03  0.004825864  0.38847026  0.095382102 -4.852087e-01
N -1.033679e-03 -3.308257e-03 -0.0014869916  0.0011950672 -6.403041e-03 -0.0047481981 -0.0090857791 -3.165633e-01 -0.109896382  0.06077214 -0.858276753 -1.179570e-01
O -7.435837e-01 -2.737301e-01 -0.2931561888  0.0624532403 -7.351970e-02  0.0165813972  0.0107652654  8.034087e-03 -0.003647926 -0.32700124 -0.073697374  3.983436e-01
P -1.887610e-02 -8.281844e-03  0.0120953610 -0.0055112850  9.161067e-03 -0.7081026816 -0.0215197071 -6.568586e-01  0.147450699 -0.03377019  0.186836647  2.342964e-02
Q -5.112824e-01 -3.158753e-01  0.5507435665  0.0047827881  1.207506e-01  0.0237725778  0.0255365539 -3.784083e-03  0.009165146  0.35232174  0.079283799 -4.270666e-01
R -1.580017e-02 -5.545668e-03  0.0095338169 -0.0054161090  1.634094e-02 -0.7047010118  0.0077147106  6.623119e-01 -0.140628767  0.02876676 -0.172816683 -1.918284e-02
         PC13        PC14        PC15         PC16          PC17          PC18
A -0.13144587 -0.08098070 -0.02454076  0.005024232  0.0068548628 -0.0087163304
B  0.22489236 -0.06594585  0.01655057 -0.723954644 -0.1793772902 -0.2019050547
C  0.06350182  0.04249943  0.01290563 -0.002550700 -0.0029193380  0.0028313002
D  0.13625476 -0.18425319  0.21476143  0.595999291 -0.0422636595  0.1959741840
E -0.15526624 -0.11547560 -0.04177224  0.002490795  0.0073294659 -0.0059979130
F -0.06090704 -0.01877042  0.02292072  0.274529691  0.0080100137 -0.9578691426
G  0.03395061  0.42286852  0.29340827  0.001678223 -0.0481454837 -0.0044208407
H  0.81614494 -0.17197121  0.15349452  0.035460715  0.1943202854 -0.0529466757
I -0.03138874 -0.40906158 -0.28385963 -0.003310920  0.0445140567  0.0033739474
J -0.12581248 -0.37763772  0.45723299 -0.017328250 -0.7385870070  0.0055287452
K  0.04011712  0.53449469  0.37638339  0.001563203 -0.0662633365 -0.0049186038
L -0.32965389 -0.30625160  0.60501298 -0.206034590  0.6111525072 -0.0103873038
M -0.01077729 -0.06532134 -0.06058002 -0.015428480  0.0009911860  0.0077789306
N  0.26634321 -0.17244991  0.18012991  0.010423377  0.0082944518  0.0008488798
O  0.01182740  0.05142697  0.04918092  0.012848866  0.0001328309 -0.0057525674
P -0.07705292  0.02894234 -0.02557832  0.010912012 -0.0091597356  0.0027707527
Q -0.01330507 -0.05801475 -0.05504574 -0.012494298  0.0008232983  0.0065952811
R  0.10040449 -0.03144543  0.03677366 -0.021420805  0.0155227342 -0.0072646567
```

From the analysis, we can see that PC1 contributes to highest variation (41%) followed by PC2 and so on. Overall, first
three PC compoenent add to ~77% of total variance and can summarize all 18 variables.

```r
> fviz_eig(pca)
```
![screenshot](https://reneshbedre.github.io//myfiles/pca_blog_screeplot.png)

*<span style="color:#696969">Fig 1: Screeplot represents eigenvalues of PCs and percentages of variation contributed by
each PC. Top principal component contribute maximum (~77%) of the total variance (only 10 PCs shown)</span>*

Now, plot the variables along the top three PCs.

```r
pca_rot = data.frame(pca$rotation)
# plot 3D
scatter3D(pca_rot$PC1, pca_rot$PC2, pca_rot$PC3, bty = "g", phi=0, xlab = "PC1", ylab ="PC2", zlab = "PC3",
            pch = 20, cex = 2, ticktype = "detailed", colvar=pca_rot$PC1, col = ramp.col(c("gold", "black", "blue")),
            colkey = list(side = 4, length = 0.5))
# add label to variables
text3D(pca_rot$PC1, pca_rot$PC2, pca_rot$PC3,  labels = rownames(pca_rot), add = TRUE, colkey = FALSE, cex = 1)
```
![screenshot](https://reneshbedre.github.io//myfiles/pca_3d.png)

*<span style="color:#696969">Fig 2: Scatterplot of variables (A to R) with respect to top three PCs, which explains
maximum (~77%) of the total variance. Color scale represents the rotation of PC1. </span>*

Interpretation: From above PCA analysis, it can be concluded that all variables except M, O, and Q has similar gene expression
profiles and form one cluster. Variables M, O and Q has high gene expression changes with respect to other variables
and form an individual cluster. From the 18 variables, it will be important to focus on M, O and Q variables.

**R and package version information used in the analysis**
```r
> sessionInfo()
R version 3.3.1 (2016-06-21)
Platform: x86_64-w64-mingw32/x64 (64-bit)
Running under: Windows 7 x64 (build 7601) Service Pack 1

locale:
[1] LC_COLLATE=English_United States.1252  LC_CTYPE=English_United States.1252    LC_MONETARY=English_United States.1252
[4] LC_NUMERIC=C                           LC_TIME=English_United States.1252

attached base packages:
[1] grid      parallel  stats     graphics  grDevices utils     datasets  methods   base

other attached packages:
 [1] plot3D_1.1.1        factoextra_1.0.5    ggplot2_2.2.1       knitr_1.17          kableExtra_0.7.0
 [6] gplots_3.0.1        Vennerable_3.0      xtable_1.8-2        gtools_3.5.0        reshape_0.8.6
[11] RColorBrewer_1.1-2  RBGL_1.50.0         graph_1.52.0        SeqGSEA_1.14.0      DESeq_1.26.0
[16] lattice_0.20-34     locfit_1.5-9.1      doParallel_1.0.10   iterators_1.0.8     foreach_1.4.3
[21] Biobase_2.34.0      BiocGenerics_0.20.0

loaded via a namespace (and not attached):
 [1] ggrepel_0.7.0        Rcpp_0.12.7          assertthat_0.1       rprojroot_1.3-2      digest_0.6.12
 [6] R6_2.2.0             plyr_1.8.4           backports_1.1.2      stats4_3.3.1         RSQLite_1.0.0
[11] evaluate_0.10.1      httr_1.2.1           highr_0.6            pillar_1.1.0         rlang_0.1.6
[16] misc3d_0.8-4         lazyeval_0.2.0       annotate_1.52.0      gdata_2.17.0         S4Vectors_0.12.0
[21] Matrix_1.2-7.1       rmarkdown_1.9        labeling_0.3         splines_3.3.1        readr_1.1.1
[26] geneplotter_1.52.0   stringr_1.3.0        RCurl_1.95-4.8       biomaRt_2.30.0       munsell_0.4.3
[31] pkgconfig_2.0.1      htmltools_0.3.6      tibble_1.4.2         IRanges_2.8.1        codetools_0.2-15
[36] XML_3.98-1.4         viridisLite_0.2.0    dplyr_0.7.4          ggpubr_0.1.6         bitops_1.0-6
[41] gtable_0.2.0         DBI_0.8              magrittr_1.5         scales_0.4.1         KernSmooth_2.23-15
[46] stringi_1.1.7        genefilter_1.56.0    bindrcpp_0.2         xml2_1.2.0           tools_3.3.1
[51] glue_1.2.0           hms_0.4.2            survival_2.39-5      AnnotationDbi_1.36.0 colorspace_1.3-2
[56] caTools_1.17.1       rvest_0.3.2          bindr_0.1
```
**<span style="color:#33a8ff">How to cite?</span>**

Bedre, R. “Principal component analysis (PCA) analysis and visualization of gene expression data using R” Renesh Bedre 
(blog), March 30, 2018, 
https://reneshbedre.github.io/blog/pca_3d.html.

<span style="color:#9e9696">If you have any questions, comments or recommendations, please email me at 
<b>reneshbe@gmail.com</b></span>

<span style="color:#9e9696"><i> Last updated: February 26, 2019</i> </span>

-->

<p>
<strong style="color:#33a8ff">How to cite?</strong>
<br />
Renesh Bedre. (2020, March 5). Bioinformatics data analysis and visualization toolkit. Zenodo. <a href="http://doi.org/10.5281/zenodo.3698145" target="_blank">http://doi.org/10.5281/zenodo.3698145</a>
<br /><br />
<span style="color:#9e9696">If you have any questions, comments or recommendations, please email me at
<b>reneshbe@gmail.com</b></span>

</p>

<p>
<link rel="stylesheet" href="https://maxcdn.bootstrapcdn.com/font-awesome/4.5.0/css/font-awesome.min.css" />

<style>
    .share-box a {
  display: inline-block;
  -webkit-box-shadow: 0 0 1px #777;
  box-shadow: 0 0 1px #777;
  padding: 5px 12px;
  margin-right: 5px;
  margin-bottom: 5px;
  text-decoration: none; }
  .share-box a:hover {
    text-decoration: none;
    -webkit-transition: background-color 200ms linear;
    -ms-transition: background-color 200ms linear;
    transition: background-color 200ms linear; }

.f {
  color: #3b5998; }
  .f:hover {
    color: #fff;
    background-color: #3b5998; }

.t {
  color: #4099FF; }
  .t:hover {
    color: #fff;
    background-color: #4099FF; }

.g {
  color: #d34836; }
  .g:hover {
    color: #fff;
    background-color: #d34836; }

.r {
  color: #ff5700; }
  .r:hover {
    color: #fff;
    background-color: #ff5700; }

.l {
  color: #0077b5; }
  .l:hover {
    color: #fff;
    background-color: #0077b5; }

.e {
  color: #444444; }
  .e:hover {
    color: #fff;
    background-color: #444444; }
</style>

<div class="share-box">
<!--<h5>Share this:</h5>-->
<a class="f" href="https://www.facebook.com/sharer/sharer.php?u=https://reneshbedre.github.io//blog/pca_3d.html" onclick="window.open(this.href, 'mywin',
'left=20,top=20,width=500,height=500,toolbar=1,resizable=0'); return false;"><i class="fab fa-facebook-f"></i></a>

<a class="t" href="https://twitter.com/intent/tweet?text=&amp;url=https://reneshbedre.github.io//blog/pca_3d.html" onclick="window.open(this.href, 'mywin',
'left=20,top=20,width=500,height=500,toolbar=1,resizable=0'); return false;"><i class="fab fa-twitter"></i></a>

  <a class="g" href="https://plus.google.com/share?url=https://reneshbedre.github.io//blog/pca_3d.html" onclick="window.open(this.href, 'mywin',
'left=20,top=20,width=500,height=500,toolbar=1,resizable=0'); return false;"><i class="fab fa-google-plus-square"></i></a>


<a class="r" href="http://www.reddit.com/submit?url=https://reneshbedre.github.io//blog/pca_3d.html" onclick="window.open(this.href, 'mywin',
'left=20,top=20,width=900,height=500,toolbar=1,resizable=0'); return false;"><i class="fab fa-reddit"></i></a>

<a class="l" href="https://www.linkedin.com/shareArticle?mini=true&amp;url=https://reneshbedre.github.io//blog/pca_3d.html" onclick="window.open(this.href, 'mywin',
'left=20,top=20,width=500,height=500,toolbar=1,resizable=0'); return false;"><i class="fab fa-linkedin-in"></i></a>

<a class="e" href="mailto:?subject=&amp;body=Check out this site https://reneshbedre.github.io//blog/pca_3d.html"><i class="fas fa-envelope-square"></i></a>

  <!--
    <a href="https://twitter.com/renbedre?ref_src=twsrc%5Etfw" class="twitter-follow-button" data-show-count="false">Follow @renbedre</a><script async src="https://platform.twitter.com/widgets.js" charset="utf-8"></script>
-->

</div>
</p>

<p><span style="color:#9e9696"><i> Last updated: March 21, 2019</i> </span></p>


        
      </section>




      <footer class="page__meta">
        
        


        <!-- commented to remove updated in footer section
        
          <p class="page__date"><strong><i class="fas fa-fw fa-calendar-alt" aria-hidden="true"></i> Updated:</strong> <time datetime="2018-03-30T01:18:08-05:00">March 30, 2018</time></p>
        
        -->
      </footer>

      <section class="page__share">
  
    <h4 class="page__share-title">Share on</h4>
  

  <a href="https://twitter.com/intent/tweet?via=renbedre&text=Principal+component+analysis+%28PCA%29+analysis+and+visualization+using+Python%20https%3A%2F%2Freneshbedre.github.io%2F%2Fblog%2Fpca_3d.html" class="btn btn--twitter" onclick="window.open(this.href, 'window', 'left=20,top=20,width=500,height=500,toolbar=1,resizable=0'); return false;" title="Share on Twitter"><i class="fab fa-fw fa-twitter" aria-hidden="true"></i><span> Twitter</span></a>

  <a href="https://www.facebook.com/sharer/sharer.php?u=https%3A%2F%2Freneshbedre.github.io%2F%2Fblog%2Fpca_3d.html" class="btn btn--facebook" onclick="window.open(this.href, 'window', 'left=20,top=20,width=500,height=500,toolbar=1,resizable=0'); return false;" title="Share on Facebook"><i class="fab fa-fw fa-facebook" aria-hidden="true"></i><span> Facebook</span></a>

  <a href="https://www.linkedin.com/shareArticle?mini=true&url=https%3A%2F%2Freneshbedre.github.io%2F%2Fblog%2Fpca_3d.html" class="btn btn--linkedin" onclick="window.open(this.href, 'window', 'left=20,top=20,width=500,height=500,toolbar=1,resizable=0'); return false;" title="Share on LinkedIn"><i class="fab fa-fw fa-linkedin" aria-hidden="true"></i><span> LinkedIn</span></a>
</section>


      
  <nav class="pagination">
    
      <a href="/blog/ecuador_workshop.html" class="pagination--pager" title="Advanced Bioinformatics Workshop
">Previous</a>
    
    
      <a href="/blog/anova.html" class="pagination--pager" title="ANOVA using Python
">Next</a>
    
  </nav>

    </div>

    
  </article>

  
  
    <div class="page__related">
      <h4 class="page__related-title">You may also enjoy</h4>
      <div class="grid__wrapper">
        
          



<div class="grid__item">
  <article class="archive__item" itemscope itemtype="https://schema.org/CreativeWork">
    
    <h2 class="archive__item-title" itemprop="headline">
      
        <a href="/blog/howtoinstall.html" rel="permalink">How to install bioinfokit
</a>
      
    </h2>
    
      <p class="page__meta"><i class="far fa-clock" aria-hidden="true"></i> 




  less than 1 minute read

</p>
    
    <p class="archive__item-excerpt" itemprop="description">










  








  




</p>
  </article>
</div>

        
          



<div class="grid__item">
  <article class="archive__item" itemscope itemtype="https://schema.org/CreativeWork">
    
    <h2 class="archive__item-title" itemprop="headline">
      
        <a href="/blog/fqqualfmt.html" rel="permalink">FASTQ Sequence Quality Format
</a>
      
    </h2>
    
      <p class="page__meta"><i class="far fa-clock" aria-hidden="true"></i> 




  3 minute read

</p>
    
    <p class="archive__item-excerpt" itemprop="description">

</p>
  </article>
</div>

        
          



<div class="grid__item">
  <article class="archive__item" itemscope itemtype="https://schema.org/CreativeWork">
    
    <h2 class="archive__item-title" itemprop="headline">
      
        <a href="/blog/bardot.html" rel="permalink">Bar graph with overlapping dots (Bar-dot) plot
</a>
      
    </h2>
    
      <p class="page__meta"><i class="far fa-clock" aria-hidden="true"></i> 




  1 minute read

</p>
    
    <p class="archive__item-excerpt" itemprop="description">










  








  




</p>
  </article>
</div>

        
          



<div class="grid__item">
  <article class="archive__item" itemscope itemtype="https://schema.org/CreativeWork">
    
    <h2 class="archive__item-title" itemprop="headline">
      
        <a href="/blog/manhat.html" rel="permalink">Manhattan plot
</a>
      
    </h2>
    
      <p class="page__meta"><i class="far fa-clock" aria-hidden="true"></i> 




  2 minute read

</p>
    
    <p class="archive__item-excerpt" itemprop="description">










  








  




</p>
  </article>
</div>

        
      </div>
    </div>
  
  
</div>

    </div>

    

    <div id="footer" class="page__footer">
      <footer>
        <!-- start custom footer snippets -->

<!-- end custom footer snippets -->
        <link rel="stylesheet" href="https://cdn.rawgit.com/jpswalsh/academicons/master/css/academicons.min.css">

<div class="page__footer-follow">
  <ul class="social-icons">
    
      <li><strong>Follow:</strong></li>
    

    
      
        
          <li><a href="https://twitter.com/renbedre" rel="nofollow noopener noreferrer"><i class="fab fa-fw fa-twitter-square" aria-hidden="true"></i> Twitter</a></li>
        
      
        
      
        
          <li><a href="https://github.com/reneshbedre" rel="nofollow noopener noreferrer"><i class="fab fa-fw fa-github" aria-hidden="true"></i> GitHub</a></li>
        
      
        
      
        
      
        
      
        
          <li><a href="https://scholar.google.com/citations?user=6pv2AiMAAAAJ&hl=en" rel="nofollow noopener noreferrer"><i class="ai ai-google-scholar-square" aria-hidden="true"></i> Google Scholar</a></li>
        
      
        
          <li><a href="https://scholar.google.com/citations?user=6pv2AiMAAAAJ&hl=en" rel="nofollow noopener noreferrer"><i class="ai ai-researchgate-square" aria-hidden="true"></i> ResearchGate</a></li>
        
      
    

    <li><a href="/feed.xml"><i class="fas fa-fw fa-rss-square" aria-hidden="true"></i> Feed</a></li>
  </ul>
</div>

<div class="page__footer-copyright">&copy; 2020 Renesh Bedre. Powered by <a href="https://jekyllrb.com" rel="nofollow">Jekyll</a> &amp; <a href="https://mademistakes.com/work/minimal-mistakes-jekyll-theme/" rel="nofollow">Minimal Mistakes</a>.</div>

      </footer>
    </div>

    
  <script src="/assets/js/main.min.js"></script>
  <script src="https://kit.fontawesome.com/4eee35f757.js"></script>







  <!-- Global site tag (gtag.js) - Google Analytics -->
<script async src="https://www.googletagmanager.com/gtag/js?id=UA-115112853-1"></script>
<script>
  window.dataLayer = window.dataLayer || [];
  function gtag(){dataLayer.push(arguments);}
  gtag('js', new Date());

  gtag('config', 'UA-115112853-1', { 'anonymize_ip': false});
</script>


<!-- Begin Inspectlet Asynchronous Code -->
    <script type="text/javascript">
    (function() {
    window.__insp = window.__insp || [];
    __insp.push(['wid', 814107329]);
    var ldinsp = function(){
    if(typeof window.__inspld != "undefined") return; window.__inspld = 1; var insp = document.createElement('script'); insp.type = 'text/javascript'; insp.async = true; insp.id = "inspsync"; insp.src = ('https:' == document.location.protocol ? 'https' : 'http') + '://cdn.inspectlet.com/inspectlet.js?wid=814107329&r=' + Math.floor(new Date().getTime()/3600000); var x = document.getElementsByTagName('script')[0]; x.parentNode.insertBefore(insp, x); };
    setTimeout(ldinsp, 0);
    })();
    </script>
    <!-- End Inspectlet Asynchronous Code -->









  </body>
</html>
